---
title: "Feeding ecology has shaped the evolution of modern sharks"
author: |
  | Mohamad Bazzi, Nicolás E. Campione, Benjamin P. Kear, Per E. Ahlberg
  | Department of Organismal Biology, Uppsala University.
  | Paleontological Institute and Museum, University of Zurich.
  |
  | Mohamad.Bazzi@ebc.uu.se
output:
  pdf_document: default
  html_document:
    df_print: paged
abstract: "Sharks are iconic predators in today’s oceans, yet their modern diversity has ancient origins. In particular, present hypotheses suggest that a combination of mass extinction, global climate change, and competition has regulated the community structure of dominant mackerel (Lamniformes) and ground (Carcharhiniformes) sharks over the last 66 million years. However, while these scenarios advocate an interplay of major abiotic and biotic events, the precise drivers remain obscure. Here, we focus on the role of feeding ecology using a geometric morphometric analysis of 3,837 fossil and extant shark teeth. Our results reveal that morphological segregation rather than competition has characterized lamniform and carcharhiniform evolution. Moreover, although lamniforms suffered a long-term disparity decline potentially linked to dietary ‘specialization’, their recent disparity rivals that of ‘generalist’ carcharhiniforms. We further confirm that low eustatic sea levels impacted lamniform disparity across the end-Cretaceous mass extinction. Adaptations to changing prey availability and the proliferation of coral reef habitats during the Paleogene also likely facilitated carcharhiniform dispersals and cladogenesis, underpinning their current taxonomic dominance. Ultimately, we posit that trophic partitioning and resource utilization shaped past shark ecology and represent critical determinants for their future species survivorship."
editor_options:
  chunk_output_type: console
---

Code compiled and maintained by Mohamad Bazzi.

**Contact**: Mohamad.Bazzi@uzh.ch; Mohammed_Bazzi@hotmail.com

Compiled: 2020-04-12

Last updated: 2021-11-22

All analyses were done using R version 4.1.1 (R Core Team 2021).
This document is an annotated version of the R code used to run geometric morphometric, phylogenetic comparative, and ecomorphological analyses.

The code is written mostly in base R syntax, but elements of tidyverse grammar have also been integrated.
Overall graphic visualization relied on the *ggplot2* package.

**The raw data supporting the conclusions of this article will be made available by the authors, without undue reservation.**

**Main reference**: Bazzi et al., 2021, Current Biology 31, 1–11.

***

# Setup
```{r}
options(width=60)
knitr::opts_chunk$set(eval = FALSE,echo=TRUE,tidy=T,fig.keep="none",tidy.opts=list(width.cutoff=60))
```

### Functions
```{r}
# Re-sample and make points equidistant.
source(file = "Functions/read points function.R")
# Confidence and prediction intervals.
source(file = "Functions/Confidence and Prediction Intervals.r")
# Function to compute Procrustes variance, rarefaction, and bootstrap statistic.
source(file = 'Functions/Morphological Disparity with Bootstrap.R')
# Custom GLS function.
source(file = "GLS Models.R")
# Function to compute the modal value.
estimate_mode <- function(x) {
  d <- hist(x, plot = FALSE)
  mode <- d$mids[which.max(d$counts)]
  return(mode)
}
# Disparity with permutation.
disparity.calc <- function(gpa, data, ages) {
  nas <- which(is.na(data[,ages]))
  if(length(nas) > 0) gm.data <- geomorph.data.frame(coords = gpa$coords[,,rownames(data)[-nas]],ages = data[-nas,ages])
  else gm.data <- geomorph.data.frame(coords = gpa$coords[,,rownames(data)],ages = data[,ages])
  disp <- morphol.disparity(coords~ages,groups = ~ages,iter = 999, data = gm.data)
  return(disp)
}
# Function for renaming tips.
rename.tips <- function(tree,names) {
    tree$tip.label <- names
    return(tree)
}
# Function for keeping tips.
keep.tip <- function(tree,tip) {
 drop.tip(tree,setdiff(tree$tip.label,tip))
}
# Function to simulate random sample distributions with replacement.
prob.stats <- function(x,n) {
  sample.x <- sample(x,n, replace = T)
  data.frame(sample.size = n, 
             sample.mean = mean(sample.x), 
             sample.sd = sd(sample.x))
}
```

### Required Libraries
```{r message=FALSE, warning=FALSE, echo=T, results='hide'}
packages <- c("geomorph","RRPP","abind","ggplot2","cowplot","reshape2","ordiBreadth",
              "tidyverse","plyr","nlme","xlsx","gridExtra","knitr","Momocs",
              "grid","MuMIn","qpcR","colorspace","phytools","sp","viridis",
              "geiger","colorspace","vegan","cluster","caper","akima",
              "Hmisc","scatterpie","ggrepel","phangorn","RColorBrewer","formatR",
              "PCDimension","ggpubr")

lapply(X = packages, FUN = library, character.only = TRUE)
```

# Data Wrangling
- Import global shark tooth data frame and index by select variables.
```{r}
Data <- read.csv2(file = "Metadata/Morphometric Tooth Database.csv",header = T)
# 1. Select first all '$Used' specimens for GM-analysis.
Data <- subset(x = Data, subset = Data$Used == "x")
# 2. Next subset by orders.
Data <- Data[Data$Order == "Lamniformes" | Data$Order == "Carcharhiniformes", ]
Data$Order <- factor(Data$Order)
# 3. Then by preservation state: i.e. completeness.
Data <- droplevels(Data[which(Data$Preservation == "C"),])
# 4. Drop levels for tps.files.
TPS.Files <- which(Data$TPS.File == "tpsKPg" | Data$TPS.File == "tpsPETM" | Data$TPS.File == "LMs_LowerTeeth")
Data <- Data[TPS.Files,]
```

### Time binning scheme
```{r}
TimeBins <- which(Data$Age.2 == "Campanian"|Data$Age.2 == "Maastrichtian"|
                  Data$Age.2 == "Danian/Selandian"|Data$Age.2 == "Thanetian"|
                  Data$Age.2 == "Ypresian"|Data$Age.2 == "Lutetian"|
                  Data$Age.2 == "Bartonian"|Data$Age.2 == "Priabonian"|
                  Data$Age.2 == "Rupelian"|Data$Age.2 == "Chattian"|
                  Data$Age.2 == "Aquitanian/Burdigalian"|Data$Age.2 == "Langhian"|
                  Data$Age.2 == "Serravallian/Tortonian"|Data$Age.2 == "Messinian"|
                  Data$Age.2 == "Zanclean"|Data$Age.2 == "Piacenzian"|
                  Data$Age.2 == "Pleistocene"|Data$Age.2 == "Recent")

Data <- Data[TimeBins,]
# Create rownames for data frame.
rownames(Data) <- paste(Data$File.Name,Data$File.Type,sep = "")
```

### Landmark data
- Raw coordinate data arranged in a .tps-file format. Points in each configuration are standardized and made equidistant using the **read2Dtps.noLMs** function. The resulting landmark scheme include 160 points.
```{r message=FALSE}
# Read tps.file: fossil raw-landmark data.
LMs <- read2Dtps.noLMs(file = "TPS Files/SharkPoints.TPS",ncurve = 2,divide.curve = TRUE, 
                       curve.1.pts = 79,curve.2.pts = 81)
# Read extant lower tooth landmark data.
Extant.Teeth <- read2Dtps.noLMs(file = "TPS Files/LMs_LowerTeeth.TPS",ncurve = 2,divide.curve = TRUE,
                                curve.1.pts = 80, curve.2.pts = 80)
# 1. Problematic specimens that should not be flipped.
except <- c(40,56,111,155,156,157,158,176,182,183,
            184,185,186,187,188,202,204,205,206,
            207,208,211,251,265,478,505,506,
            584,590,591,592,593,595,609,704,
            933,934,935,936,937)
# 2. Flip orientation of extant teeth to match those of fossil specimens.
Extant.Teeth$coords[,1,-except] <- Extant.Teeth$coords[,1,-except]*(-1)
# 3. Reshuffle the sequence of points for consistency in point-ordering between curves.
Extant.Teeth$coords[,,-except] <- Extant.Teeth$coords[c(81:160,80:1),,-except]
Extant.Teeth$coords[,,except] <- Extant.Teeth$coords[c(1:80,160:81),,except]
```

### Combine multi-dimensional arrays
```{r message=FALSE,echo=TRUE,results='hide'}
# Set difference of subsets.
setdiff(x = rownames(Data),y = dimnames(LMs$coords)[[3]])
# Intersection of subsets.
FossilNames <- intersect(x = rownames(Data),y = dimnames(LMs$coords)[[3]])
ExtantNames <- intersect(x = rownames(Data),y = dimnames(Extant.Teeth$coords)[[3]])
# Combine landmark-coordinate datasets and match with rownames in the data frame.
TotalLandmarks <- abind(LMs$coords[,,FossilNames],Extant.Teeth$coords[,,ExtantNames])
setdiff(x = rownames(Data),y = dimnames(TotalLandmarks)[[3]])
dim(TotalLandmarks) # 3837 specimens.
```

### Spatio-temporal display of shark teeth
- Scatter pie plots depict the total sample size of shark clades and their proportional representation through time.
```{r warning=FALSE}
# Read in data frame.
Map.Data <- read.xlsx(file = "Metadata/Geographic Graph.xlsx",sheetIndex = 1,sheetName = "Lamniformes",header = T)
Map.Data <- Map.Data[-1]
# Define the basic characteristics of the plot.
world <- map_data('world')
world <- fortify(world)
p <- ggplot() + theme(axis.text.y = element_text(size = 7),
                      axis.text.x = element_text(size = 7))
p <- p + geom_polygon(data = world,aes(x = long, y = lat, group = group),
                      alpha = .5,colour = "lightgray", fill = "grey90",size = .25) + coord_quickmap(expand = T)
# Plot.
p <- p + geom_scatterpie(aes(x = lon, y = lat, group = region,r = sqrt(totalSample)/2),
                    data = Map.Data, cols = LETTERS[1:8],
                    alpha = .5,sorted_by_radius = T) +
  theme_bw() +
  theme(legend.position = "top") +
  geom_text_repel(data = Map.Data, aes(x = lon, y = lat, label = region), size = 2,
                  box.padding = unit(0.35, "lines"),
                  point.padding = unit(0.3, "lines"))
print(p)
```

# Analysis
- The superimposition of landmark configurations was achieved using a partial GPA (Dryden & Mardia 201). The difference between a partial and full GPA is with respect to the scaling step. For the partial variant, the consensus shape (i.e., mean) is re-scaled to have unit centroid size of 1.0.
- Further, we used a bending energy criterion to minimize the sum-of-squares between adjacent sliding semi-landmarks.
- Here, only the first three eigenvectors (variance > 10) were retained for downstream analyses and interpretation.
- The biological importance of lower dimensions was assessed by examining Thin plate spline deformation grids.
```{r echo=TRUE,results='hide'}
# Generalized Procrustes Analysis (GPA)
GPA <- gpagen(A = TotalLandmarks,curves = as.matrix(LMs$sliders[-79,]),ProcD = FALSE,print.progress = TRUE,max.iter = NULL)
summary(GPA)
gpa.file <- saveRDS(object = GPA, file = "gpa.file.rds")
```

```{r echo=TRUE,results='hide'}
# Principal Component Analysis (PCA)
PCA <- gm.prcomp(A = GPA$coords)
PCA %>% summary()
# Proportion of Variance.
var_exp <- PCA$sdev^2/sum(PCA$sdev^2)
names(var_exp) <- seq(1:length(var_exp))
round(var_exp[1:4]*100,2)
# Cumulative Proportion
cum_pro <- cumsum(PCA$sdev^2/sum(PCA$sdev^2))
```

### Visualize PCA-results
- Scree plot showing the distribution of eigenvalues.
- Cumulative proportion (i.e., the accumulated amount of explained variance).
- The first 3 components explain ~86% of the variance.
```{r warning=FALSE}
layout.matrix <- matrix(c(1,2), nrow = 1, ncol = 2,byrow = FALSE)
layout(mat = layout.matrix,heights = c(6,6),widths = c(5,3),respect = TRUE)
# Proportion of Variance.
barplot(var_exp[1:10],space = .05,beside = TRUE,horiz = FALSE,
        main = expression(paste("Distribution of eigenvalues" ~ (lambda[k]))),
        font.main = 4,las = 1,xpd = FALSE,cex.axis = .6 ,cex.names = .6,
        cex.lab = .6, xlab = "Principal Components",
        ylab = "Percentage of variances",
        col = "grey70", border = "black",font.lab = 2)
# Add connected line segments to the plot.
lines(var_exp[1:10],type = "b", pch = 19, col = scales::alpha("orange",.5), cex = 1.5)
abline(h = mean(var_exp[1:10]), lwd = 2, lty = "dashed", col = "black")
# Cumulative Proportion.
plot(cum_pro[1:10],type = "o", pch = 21,
     main = "Cumulative Proportion", lty = 1, lwd = 2,cex.axis = .6,
     cex.lab = .6,xlab = "Principal Components", ylab = "",
     col = "black", bg = "grey70", font.lab = 2, bty = "n")

dev.off()
```

```{r}
# The Broken Stick Method
bsDimension(var_exp, FUZZ = 0.05)
dd <- c(dim(GPA$coords)[3], dim(GPA$coords)[1])
# Estimating Number of Principal Components Using the Auer-Gervini Method
aG <- AuerGervini(var_exp, dd=dd, epsilon = 2e-16)
summary(aG)
```

### Shark tooth morphospace
```{r warning=FALSE,echo=T}
# Data frame.
shape_space <- data.frame(pc1 = PCA$x[,1],
                          pc2 = PCA$x[,2],
                          clades = factor(Data$Order))
# PCA graph.
ggplot(shape_space, aes(x = pc1, y = pc2, color = clades)) +
  geom_point(size = 1.5) +
  geom_vline(xintercept = 0, lwd = 1, linetype = "solid", col = "black") +
  geom_hline(yintercept = 0, lwd = 1, linetype = "solid", col = "black") +
  labs(x = "Principal Component 1 (64.79)", y = "Principal Component 2 (11.43%)") +
  theme_classic() +
  theme(legend.position = "none") +
  theme(axis.title = element_text(size = 7),
        axis.text.y = element_text(angle = 90, size = 5),
        axis.text.x = element_text(size = 5)) +
  theme(axis.line = element_line(size = .5, linetype = "solid")) +
  scale_color_manual(values = c("#377EB8","#E84646")) +
  coord_fixed()
```

### Thin Plate Spline Deformation Grids
- Visualize shape variation along each axis.
- The shape at specific positions along our main axes of variation are computed.
- Here we used the **tps_grid** function from the Momocs package to plot deformation grids.
- In addition to PC1-PC3 we also visualize shape variability along PC4 and PC5. This was done using the **plotRefToTarget** function in geomorph.
```{r warning=FALSE,eval=FALSE}
# Shape variation on PC1.
pc1_shape_a <- GPA$consensus + matrix(-0.6*(PCA$rotation[,1]),byrow=T,160,2)
pc1_shape_b <- GPA$consensus + matrix(-0.3*(PCA$rotation[,1]),byrow=T,160,2)
pc1_shape_c <- GPA$consensus + matrix(0.0*(PCA$rotation[,1]),byrow=T,160,2)
pc1_shape_d <- GPA$consensus + matrix(0.3*(PCA$rotation[,1]),byrow=T,160,2)
pc1_shape_e <- GPA$consensus + matrix(0.6*(PCA$rotation[,1]),byrow=T,160,2)

# Plot TPS-grids.
par(mfrow = c(1,5))
tps_grid(pc1_shape_a,pc1_shape_a,grid.size = 10,shp = T,legend = T,shp.lwd = 6)
tps_grid(pc1_shape_b,pc1_shape_b,grid.size = 10,shp = T,legend = T,shp.lwd = 6)
tps_grid(pc1_shape_c,pc1_shape_c,grid.size = 10,shp = T,legend = T,shp.lwd = 6)
tps_grid(pc1_shape_d,pc1_shape_d,grid.size = 10,shp = T,legend = T,shp.lwd = 6)
tps_grid(pc1_shape_e,pc1_shape_e,grid.size = 10,shp = T,legend = T,shp.lwd = 6)
dev.off()

# Shape variation on PC2.
pc2_shape_a <- GPA$consensus + matrix(-0.4*(PCA$rotation[,2]),byrow=T,160,2)
pc2_shape_b <- GPA$consensus + matrix(-0.2*(PCA$rotation[,2]),byrow=T,160,2)
pc2_shape_c <- GPA$consensus + matrix(0.0*(PCA$rotation[,2]),byrow=T,160,2)
pc2_shape_d <- GPA$consensus + matrix(0.2*(PCA$rotation[,2]),byrow=T,160,2)
pc2_shape_e <- GPA$consensus + matrix(0.4*(PCA$rotation[,2]),byrow=T,160,2)

# Plot TPS-grids.
par(mfrow = c(1,5))
tps_grid(pc2_shape_a,pc2_shape_a,grid.size = 10,shp = T,legend = T,shp.lwd = 6)
tps_grid(pc2_shape_b,pc2_shape_b,grid.size = 10,shp = T,legend = T,shp.lwd = 6)
tps_grid(pc2_shape_c,pc2_shape_c,grid.size = 10,shp = T,legend = T,shp.lwd = 6)
tps_grid(pc2_shape_d,pc2_shape_d,grid.size = 10,shp = T,legend = T,shp.lwd = 6)
tps_grid(pc2_shape_e,pc2_shape_e,grid.size = 10,shp = T,legend = T,shp.lwd = 6)
dev.off()

# Shape variation on PC3.
pc3_shape_a <- GPA$consensus + matrix(-0.3400487*(PCA$rotation[,3]),byrow=T,160,2)
pc3_shape_b <- GPA$consensus + matrix(0.2980234*(PCA$rotation[,3]),byrow=T,160,2)
# Plot TPS-grids.
par(mfrow = c(1,2))
tps_grid(pc3_shape_a,pc3_shape_a,grid.size = 10,shp = T,legend = T,shp.lwd = 6)
tps_grid(pc3_shape_b,pc3_shape_b,grid.size = 10,shp = T,legend = T,shp.lwd = 6)
dev.off()

# Extreme shapes on PC4 and PC5.
par(mfrow = c(2,2))
plotRefToTarget(M1 = GPA$consensus,M2 = PCA$shapes$shapes.comp4$min)
plotRefToTarget(M1 = GPA$consensus,M2 = PCA$shapes$shapes.comp4$max)
plotRefToTarget(M1 = GPA$consensus,M2 = PCA$shapes$shapes.comp5$min)
plotRefToTarget(M1 = GPA$consensus,M2 = PCA$shapes$shapes.comp5$max)
dev.off()
```

### Linear Model Evaluation using ANOVA
- Here we test if Lamniformes and Carcharhiniformes occupy different areas of morphospace.
- We use Type I (“sequential” sum of squares: SS) because our treatments are nearly balanced.
- Group sample sizes: Lamniformes (N=1924) and Carcharhiniformes (N=1913).
- To obtain model fitting results associated with single axes of variation simply substitute elements found in the relevant data frame.
```{r warning=FALSE,echo=TRUE,results='hide'}
# 1. Data frame.
aov.df <- rrpp.data.frame(all.axes = PCA$x,PC1 = PCA$x[,1],
                          PC2 = PCA$x[,2],PC3 = PCA$x[,3],
                          clades = factor(Data$Order))
# 2. Normality and homogeneity of variance.
layout(matrix(1:3,nrow = 1,ncol = 3,byrow = T),respect = T)
boxplot(PCA$x[,1] ~ Data$Order, ylab = "PC1", xlab = "")
boxplot(PCA$x[,2] ~ Data$Order, ylab = "PC2", xlab = "")
boxplot(PCA$x[,3] ~ Data$Order, ylab = "PC3", xlab = "Clades")
dev.off()
# 3. Fit linear models.
fit.model <- lm.rrpp(f1 = all.axes ~ clades,iter = 999,RRPP = TRUE,SS.type = "I",data = aov.df)
anova(fit.model, effect.type = "F",error = "Residuals")
```

```{r}
# Re-level ages: 18 stages.
bins <- c("Campanian","Maastrichtian","Danian/Selandian","Thanetian","Ypresian",
          "Lutetian","Bartonian","Priabonian","Rupelian","Chattian",
          "Aquitanian/Burdigalian","Langhian","Serravallian/Tortonian","Messinian",
          "Zanclean","Piacenzian","Pleistocene","Recent")
# 17 stages (excluding the Pleistocene).
bins.x <- c("Campanian","Maastrichtian","Danian/Selandian","Thanetian","Ypresian",
          "Lutetian","Bartonian","Priabonian","Rupelian","Chattian",
          "Aquitanian/Burdigalian","Langhian","Serravallian/Tortonian","Messinian",
          "Zanclean","Piacenzian","Recent")
```

### Morphospace through time
```{r warning=FALSE}
# Data frame.
morphospaceTime <- data.frame(pc1 = PCA$x[,1], pc2 = PCA$x[,2], 
                              pc3 = PCA$x[,3],groups = factor(Data$Order),
                              age = factor(Data$Age.2))
# Arrange time-bins. 
morphospaceTime$age <- factor(morphospaceTime$age,levels = c("Campanian","Maastrichtian","Danian/Selandian",
                                                             "Thanetian","Ypresian","Lutetian","Bartonian",
                                                             "Priabonian","Rupelian","Chattian",
                                                             "Aquitanian/Burdigalian","Langhian",
                                                             "Serravallian/Tortonian","Messinian",
                                                             "Zanclean","Piacenzian","Pleistocene","Recent"))
# Arrange clades.
morphospaceTime$groups <- factor(morphospaceTime$groups, c("Lamniformes", "Carcharhiniformes"))
# Stratigraphically arrange the time-bins.
# Only for the histogram plot.
morphospaceTime$age <- with(morphospaceTime,factor(morphospaceTime$age,levels = rev(levels(morphospaceTime$age))))

# Convert object into a molten data frame.
time_Melt <- melt(morphospaceTime)
# Histograms.
ggplot(time_Melt, aes(x = value, color = groups,fill = groups)) + 
  geom_histogram(binwidth = 0.02,alpha = .35, fill = "white",
                 position = "identity") + 
  facet_grid(age~variable,scales = "free") +
  theme(legend.position = "none")
```

### Morphospace plot
```{r warning=FALSE}
# Plot
Time_Plot <- filter(time_Melt, !is.na(age)) %>% 
ggplot(aes(x = age, y = value,colour = factor(groups))) +
  geom_boxplot(outlier.colour = NULL,outlier.alpha = NULL,outlier.shape = 21) +
  stat_summary(fun = "mean", geom = "point",lwd = 1,position = position_dodge(width = 0.75),bg = 'white',pch = 21) +
  stat_summary(aes(group=groups), fun=mean, geom="line",position = position_dodge(width = 0.75)) +
  scale_fill_manual(values = c("#E84646","#377EB8",.5)) +
  scale_color_manual(values = c("#E84646","#377EB8",.5)) +
  facet_grid(.~groups) +
  geom_vline(xintercept = c(2.5,4.5,8.5,10.5,14.5,16.5), lwd = 1, color = "black") +
  theme(legend.position = "top") +
  theme(axis.title = element_text(color = "#666666", face = "bold", size = 7),
        axis.text.y = element_text(angle = 90, size = 7),
        axis.text.x = element_text(size = 7)) +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1)) +
  theme(panel.spacing = unit(0, "lines")) +
  theme(aspect.ratio = 1/3) +
  facet_wrap(~ variable,nrow = 3, scales = "free_y",shrink = F)

print(Time_Plot)
```

### Heterodont Morphospace
- Evaluate temporal patterns in morphospace based on select tooth positions.
- This analysis is limited to anterior (n=609) and lateroposterior (n=1323) teeth. **NB**: anterolateral and posterolateral were treated as lateroposterior.
```{r message=FALSE}
# New column.
Data$Ne.Position <- NULL
Data$Ne.Position <- Data$Relative.Position
Data$Ne.Position <- factor(Data$Ne.Position)

# Select positions.
tooth.positions <- c("anterior","lateroposterior")
# New frame.
ex.MonoHe <- Data[Data$Ne.Position %in% tooth.positions, ]
# Remove the Pleistocene from both data frames.
ex.MonoHe <- ex.MonoHe[ex.MonoHe$Age.2 != "Pleistocene", ]
# gg.frame.
space.MonoHet <- data.frame(pc1 = PCA$x[rownames(ex.MonoHe),1],
                            pc2 = PCA$x[rownames(ex.MonoHe),2], 
                            pc3 = PCA$x[rownames(ex.MonoHe),3],groups = factor(ex.MonoHe$Order),
                            position = factor(ex.MonoHe$Ne.Position),
                            age = factor(ex.MonoHe$Age.2))
# Arrange time-bins. 
space.MonoHet$age <- factor(space.MonoHet$age,levels = bins.x)
# Arrange clades.
space.MonoHet$groups <- factor(space.MonoHet$groups, c("Lamniformes", "Carcharhiniformes"))
# Convert object into a molten data frame.
hetero_Melt <- melt(space.MonoHet)

# Plot.
Hetero_Plot <- filter(hetero_Melt, !is.na(age)) %>% 
  ggplot(aes(x = age, y = value,colour = factor(groups))) +
  geom_boxplot(outlier.colour = NULL,outlier.alpha = NULL,outlier.shape = 21) +
  scale_fill_manual(values = c("#E84646","#377EB8",.5)) +
  scale_color_manual(values = c("#E84646","#377EB8",.5)) +
  geom_vline(xintercept = c(2.5,4.5,8.5,10.5,14.5,16.5), lwd = 1, color = "black") +
  theme(legend.position = "top") +
  theme(axis.title = element_text(color = "#666666", face = "bold", size = 7),
        axis.text.y = element_text(angle = 90, size = 7),
        axis.text.x = element_text(size = 7)) +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1)) +
  theme(panel.spacing = unit(0, "lines")) +
  theme(aspect.ratio = 1/2) +
  facet_wrap(c("variable","position"),nrow = 3, scales = "free_y",shrink = F)

print(Hetero_Plot)
```

### Clade-specific data frames and pc-scores
```{r}
# Frames.
Lamniformes <- which(Data$Order == "Lamniformes")
# Lamn.Df$Order <- factor(Lamn.Df$Order)
Lamn.Df <- Data[Lamniformes, ]

Carcharhiniformes <- which(Data$Order == "Carcharhiniformes")
# Carch.Df$Order <- factor(Carch.Df$Order)
Carch.Df <- Data[Carcharhiniformes, ]

# PC-scores.
Lamn.scores <- PCA$x[rownames(Lamn.Df), ]
Carch.scores <- PCA$x[rownames(Carch.Df), ]
```

### Compare extant (Holocene) shark tooth distributions
```{r warning=FALSE}
# Data frame.
extant.distrib <- data.frame(PC1 = PCA$x[Data$Age.2 == "Recent",1],
                             PC2 = PCA$x[Data$Age.2 == "Recent",2],
                             PC3 = PCA$x[Data$Age.2 == "Recent",3],
                             groups = Data$Order[Data$Age.2 == "Recent"],
                             families = Data$Family[Data$Age.2 == "Recent"])
# Melt.
extant.distrib <- melt(extant.distrib)
# Summary statistics.
pc.mean <- ddply(.data = extant.distrib,c("variable","groups"),summarise, grp.mean=mean(value))
# Plot data.
ggplot(extant.distrib, aes(x = value, color = groups)) +
  geom_histogram(aes(fill = groups),bins = 30,alpha = 0.4,lwd = 1) +
  scale_color_manual(values = c("#377EB8","#E84646")) +
  scale_fill_manual(values = c("#377EB8","#E84646")) +
  theme(legend.position = "top",
        axis.title = element_text(color = "#666666", face = "bold", size = 7),
        axis.text.y = element_text(size = 7),
        axis.text.x = element_text(size = 7)) +
  geom_vline(data=pc.mean, aes(xintercept=grp.mean, color=groups),
             linetype="dashed",lwd = 1) +
  theme(aspect.ratio = 1) + 
  facet_wrap(. ~ variable,scales = "free",ncol = 3)
```

```{r}
# Plot extant families.
ggplot(extant.distrib, aes(x = value, color = families)) +
  geom_boxplot() +
  facet_wrap(. ~ variable,scales = "free",nrow = 3)
```

```{r}
# Variables.
X.1 <- Lamn.scores[Lamn.Df$Age.2 == "Recent",1]
Y.1 <- Carch.scores[Carch.Df$Age.2 == "Recent",1]
# K-S and Wilcoxon test.
ks.test(X.1,Y.1,exact = FALSE)
wilcox.test(X.1,Y.1,conf.int = TRUE,conf.level = .95)
# Empirical cumulative distribution function.
plot(ecdf(X.1), xlim = range(c(X.1, Y.1)), col = "black", pch = 19)
plot(ecdf(Y.1), add = TRUE, lty = "dashed", col = "Purple", pch = 19)
```

### Morphological resampling
The nature of our extant and fossil shark teeth samples substantially deviate from each other.
For this reason we apply a *randomization process* to our extant dataset by simulating multiple random samples that we compare against a reference (i.e., population) distributions. We rerun the procedure a 1000 times for 10 samples of different sizes. The overall distributions and their means are then compared.
**NB:** Make sure to detach the plyr package prior to running this chunk of code.
```{r warning=FALSE}
# Population mean of Lamniformes on PC1.
L.mean <- -0.04364368
# Population standard deviation of Lamniformes on PC1.
L.sd <- 0.2136296
# Population data frame.
pop.dist <- data.frame(scores = Lamn.scores[Lamn.Df$Age.2 == "Recent",1])

# The 'rerun' function works similar to replicate function.
# Here a total of 100 samples are generated of size x from the known population distribution.
# Lamniformes.
sample.10 <- rerun(100,prob.stats(x = Lamn.scores[Lamn.Df$Age.2 == "Recent",1],n = 10)) %>% bind_rows()
sample.20 <- rerun(100,prob.stats(x = Lamn.scores[Lamn.Df$Age.2 == "Recent",1],n = 20)) %>% bind_rows()
sample.30 <- rerun(100,prob.stats(x = Lamn.scores[Lamn.Df$Age.2 == "Recent",1],n = 30)) %>% bind_rows()
sample.40 <- rerun(100,prob.stats(x = Lamn.scores[Lamn.Df$Age.2 == "Recent",1],n = 40)) %>% bind_rows()
sample.50 <- rerun(100,prob.stats(x = Lamn.scores[Lamn.Df$Age.2 == "Recent",1],n = 50)) %>% bind_rows()
sample.60 <- rerun(100,prob.stats(x = Lamn.scores[Lamn.Df$Age.2 == "Recent",1],n = 60)) %>% bind_rows()
sample.70 <- rerun(100,prob.stats(x = Lamn.scores[Lamn.Df$Age.2 == "Recent",1],n = 70)) %>% bind_rows()
sample.80 <- rerun(100,prob.stats(x = Lamn.scores[Lamn.Df$Age.2 == "Recent",1],n = 80)) %>% bind_rows()
# Combine them all.
df.combine <- bind_rows(sample.10,sample.20,sample.30,sample.40,
                        sample.50,sample.60,sample.70,sample.80) %>% mutate(ss = as.factor(sample.size))

# Compare the sampling distributions of the means for samples of size X.
# How does sample size affect the sampling distribution of means.
ggplot(df.combine, aes(x = sample.mean,y = ..density.., fill = ss)) +
  geom_histogram(bins = 50) +
  theme(axis.title = element_text(color = "#666666", face = "bold", size = 7),
        axis.text.y = element_text(size = 7),
        axis.text.x = element_text(size = 7)) +
  geom_vline(xintercept = L.mean,linetype ="dashed") +
  facet_wrap(~ss, scales = "free_y",ncol = 4) +
  geom_density(mapping = aes(y=..density..),colour="black",alpha = .5) +
  scale_fill_brewer(palette = "PRGn") +
  labs(x = "Principal Component 1", y = "Density",
       title = "Lamniform population vs. mean sampling distributions",
       subtitle = "Black line indicate the population mean") +
  theme(panel.spacing = unit(0, "lines")) +
  theme(aspect.ratio = 1)

# Compute and save means and sd.
detach(package:plyr)

df.means <- df.combine %>% dplyr::group_by(ss) %>% 
  summarise(mean.of.means = mean(sample.mean),sd.of.means = sd(sample.mean))
```

```{r}
# Carcharhiniformes.
sample.c10 <- rerun(100,prob.stats(x = Carch.scores[Carch.Df$Age.2 == "Recent",1],n = 10)) %>% bind_rows()
sample.c20 <- rerun(100,prob.stats(x = Carch.scores[Carch.Df$Age.2 == "Recent",1],n = 20)) %>% bind_rows()
sample.c30 <- rerun(100,prob.stats(x = Carch.scores[Carch.Df$Age.2 == "Recent",1],n = 30)) %>% bind_rows()
sample.c40 <- rerun(100,prob.stats(x = Carch.scores[Carch.Df$Age.2 == "Recent",1],n = 40)) %>% bind_rows()
sample.c50 <- rerun(100,prob.stats(x = Carch.scores[Carch.Df$Age.2 == "Recent",1],n = 50)) %>% bind_rows()
sample.c60 <- rerun(100,prob.stats(x = Carch.scores[Carch.Df$Age.2 == "Recent",1],n = 60)) %>% bind_rows()
sample.c70 <- rerun(100,prob.stats(x = Carch.scores[Carch.Df$Age.2 == "Recent",1],n = 70)) %>% bind_rows()
sample.c80 <- rerun(100,prob.stats(x = Carch.scores[Carch.Df$Age.2 == "Recent",1],n = 80)) %>% bind_rows()

# Combine them all.
C.df.combine <- bind_rows(sample.c10,sample.c20,sample.c30,sample.c40,
                          sample.c50,sample.c60,sample.c70,sample.c80) %>% mutate(ss = as.factor(sample.size))

# Compare the sampling distributions of the means for samples of size X.
# How does sample size affect the sampling distribution of means?
C.mean <- 0.08424445

ggplot(C.df.combine, aes(x = sample.mean,y = ..density.., fill = ss)) +
  geom_histogram(bins = 50) +
  theme(axis.title = element_text(color = "#666666", face = "bold", size = 7),
        axis.text.y = element_text(size = 7),
        axis.text.x = element_text(size = 7)) +
  geom_vline(xintercept = C.mean,linetype ="dashed") +
  facet_wrap(~ss,scales = "free_y",ncol = 4) +
  geom_density(mapping = aes(y=..density..),colour="black",alpha = .5) +
  scale_fill_brewer(palette = "PRGn") +
  labs(x = "Principal Component 1", y = "Density",
       title = "Carcharhiniform population vs. mean sampling distributions",
       subtitle = "Black line indicate the population mean") +
  theme(panel.spacing = unit(0, "lines")) +
  theme(aspect.ratio = 1)

# Compute and save means.
df.C.means <- C.df.combine %>% dplyr::group_by(ss) %>% 
  summarise(mean.of.means = mean(sample.mean),sd.of.means = sd(sample.mean))
```

### Anova statistics applied to morphospace time-series
- Linear model assessment using a randomized residual permutation procedure (RRPP).
- P-values are FDR adjusted accounting for the unbalanced data design.
- Pairwise statistics are based on the distances between means (test.type = "dist").
- Pairwise results are summarized and provided in Excel sheet.
```{r echo=TRUE,results='hide'}
# 1. Lamniformes
lamn.rrpp <- rrpp.data.frame(all.pcs =  Lamn.scores[, ],pc1 = Lamn.scores[,1],
                             pc2 = Lamn.scores[,2], pc3 = Lamn.scores[,3],
                             gp = factor(Lamn.Df$Age.2))
# Linear model-fit.
l.aov.pc <- lm.rrpp(f1 = all.pcs~gp,iter = 999,RRPP = TRUE, SS.type = "I", data = lamn.rrpp)
# Anova table.
anova(l.aov.pc, effect.type = "F",error = "Residuals")
# Change [,1] for results from PC2 & PC3.
pw.pc <- pairwise(l.aov.pc, groups = interaction(Lamn.Df$Age.2))
pPC.adjust <- summary(pw.pc, test.type = "dist", confidence = 0.95, stat.table = TRUE)
# Column names.
c.name <- colnames(pPC.adjust$pairwise.tables$P)
# FDR-adjustment.
round(matrix(p.adjust(pPC.adjust$pairwise.tables$P, method = "fdr"),
             nrow = 18, dimnames = list(c.name,c.name)),3)
```

```{r echo=TRUE,results='hide'}
# 2. Carcharhiniformes.
carch.rrpp <- rrpp.data.frame(all.pcs =  Carch.scores[, ],pc1 = Carch.scores[,1],
                              pc2 = Carch.scores[,2], pc3 = Carch.scores[,3],
                              gp = factor(Carch.Df$Age.2))
# Linear model-fit.
c.aov.pc <- lm.rrpp(f1 = all.pcs~gp,iter = 999,RRPP = TRUE, SS.type = "I", data = carch.rrpp)
# Anova table.
print(anova(c.aov.pc, effect.type = "F"))
c.pw.pc <- pairwise(fit = c.aov.pc, groups = interaction(Carch.Df$Age.2))
pPC.adjust <- summary(c.pw.pc, test.type = "dist", confidence = 0.95, stat.table = TRUE)
# Column names.
c.name <- colnames(pPC.adjust$pairwise.tables$P)
# FDR-adjustment.
round(matrix(p.adjust(pPC.adjust$pairwise.tables$P, method = "fdr"),nrow = 18, dimnames = list(c.name,c.name)),3)
# How many (=length) and which ones are significant?
which(pPC.adjust[[1]]$P < 0.05)
```

### Procrustes variance (disparity analysis)
- Rarefaction applied to Lamniformes rely on the lowest sampled time-bin which is the Piacenzian (N=12).
- The Pleistocene (N=4) is initially omitted from the analysis, but later re-introduced.
```{r message=FALSE,warning=FALSE,cache=FALSE,eval=FALSE}
# Omit the Pleistocene.
Lamn.Df <- Lamn.Df[Lamn.Df$Age.2 != "Pleistocene", ]
Lamn.Df$Age.2 <- factor(Lamn.Df$Age.2)

# Disparity calculation with rarefaction.
Lamn.Disp <- error.plot(gpa.coords = GPA$coords[,,rownames(Lamn.Df)],
                        blank = FALSE,groups = Lamn.Df$Age.2,
                        order = c(3,8,5,15,16,7,2,11,13,4,1,6,14,9,17,10,12),
                        replicates = 999,rarefy.par = list(min.N = 12,reps = 999))
# Save output.
saveRDS(object = Lamn.Disp,file = "Lamn.ProcrustesVariance.rds")

# Re-sampling procedure.
Lamn.Perm <- disparity.calc(gpa = GPA,data = Lamn.Df,ages = "Age.2")
ages <- colnames(Lamn.Perm$PV.dist.Pval)
# Adjust p-values.
stat_res <- round(matrix(p.adjust(Lamn.Perm$PV.dist.Pval, method = "fdr"),nrow = 17,dimnames = list(ages,ages)),3)
# Procrustes variance with the Pleistocene.
Lamniformes <- which(Data$Order == "Lamniformes")
Lamn.Df <- Data[Lamniformes, ]
PV.Ple <- morphol.disparity(f1 = GPA$coords[,,rownames(Lamn.Df)] ~ Age.2,groups = ~Age.2,
                            iter = 999,data = Lamn.Df)
```

```{r message=FALSE,warning=FALSE,cache=FALSE,eval=FALSE}
# Disparity calculation.
Carch.Disp <- error.plot(gpa.coords = GPA$coords[,,rownames(Carch.Df)],
                         blank = FALSE,groups = Carch.Df$Age.2,
                         order = c(3,8,5,16,17,7,2,12,14,4,1,6,15,9,18,10,11,13),
                         replicates = 999,rarefy.par = list(min.N = 8,reps = 999))

barplot(Carch.Disp$bootstrap.results[,1],ylab = "PV", col = "blue",las = 2)
abline(h = c(0.04,0.06),lty = c(1,2))

# Save output.
saveRDS(object = Carch.Disp,file = "Carch.ProcrustesVariance.rds")

# Re-sampling procedure.
Carch.Perm <- disparity.calc(gpa = GPA,data = Carch.Df,ages = "Age.2")
# Adjust p-value.
C.ages <- colnames(Carch.Perm$PV.dist.Pval)
stat_res_Carch <- round(matrix(p.adjust(Carch.Perm$PV.dist.Pval, method = "fdr"),nrow = 18,dimnames = list(C.ages,C.ages)),3)
```

### First-order sensitivity analysis
- Reassess the Miocene-Pliocene decline for lamniform sharks by excluding otodontids from the data.
```{r message=FALSE,warning=FALSE,cache=FALSE,eval=FALSE}
# Drop level.
ex.otodontid <- Lamn.Df[Lamn.Df$Family != "Otodontidae",]
ex.otodontid$Family <- factor(ex.otodontid$Family)
# Age sample sizes
table(ex.otodontid$Age.2)
# Disparity.
PV.ex.O <- morphol.disparity(f1 = GPA$coords[,,rownames(ex.otodontid)] ~ Age.2,groups = ~Age.2,iter = 999,data = ex.otodontid)
# Plot.
plot(PV.ex.O$Procrustes.var[c(3,8,5,16,17,7,2,12,14,4,1,6,15,9,18,10,11,13)], type = "o",ylab = "PV", col = "red")
abline(v = c(11,15),lty = 1)
polygon(x = c(11,15,15),y = c(0,-15,15),col=scales::alpha("gray",.4),border=NA)
# All lamniforms.
lines(PV.Ple$Procrustes.var[c(3,8,5,16,17,7,2,12,14,4,1,6,15,9,18,10,11,13)],type = "o", col = "black")
mtext(text = "Miocene",side = 3,line = -1,at = 13,font = 2,cex = .5)
legend(x = 2,y = 0.03 ,legend = c("Lamniformes","Excluding otodontids"),
       col = c("black","red"),lty = c(19,19),cex = .5)
```

### Second-order sensitivity analysis
- Our detection of resurgent disparity among extant lamniforms is unexpected because it approaches the historical peak in Late Cretaceous fossil taxa.
- The datasets used are however simply not comparable (i.e., extant vs. fossil) and the modern rebound is therefore likely artefactual.
- We explore the disparity of extant species based on average tooth shapes and compare against previous results.
```{r message=FALSE}
# Extant Lamniformes.
extant.lamn <- Lamn.Df[Lamn.Df$Age.2 == "Recent", ]
extant.lamn$genspe <- NULL
extant.lamn$genspe <- paste(extant.lamn$Genus,extant.lamn$Species,sep = "_")
# Extant Carcharhiniformes.
extant.carch <- Carch.Df[Carch.Df$Age.2 == "Recent", ]
extant.carch$genspe <- NULL
extant.carch$genspe <- paste(extant.carch$Genus,extant.carch$Species,sep = "_")
# Average species coordinates.
L.means <- rowsum(two.d.array(GPA$coords[,,rownames(extant.lamn)]), extant.lamn$genspe)/as.vector(table(extant.lamn$genspe))
L.array <- arrayspecs(A = L.means,p = 160,k = 2)
C.means <- rowsum(two.d.array(GPA$coords[,,rownames(extant.carch)]), extant.carch$genspe)/as.vector(table(extant.carch$genspe))
C.array <- arrayspecs(A = C.means,p = 160,k = 2)
# Procrustes variance.
average.lamn <- morphol.disparity(f1 = L.array ~ 1,groups = NULL,iter = 999)
# Procrustes variance.
average.carch <- morphol.disparity(f1 = C.array ~ 1,groups = NULL,iter = 999)
```

### Third-order sensitivity analysis
Here, we investigate the effect of heterodonty as a nominal factor on shark tooth disparity.
Tooth positions and jaw units are treated as additional explanatory (predictor) variables to help explain variation in the response variable (i.e. disparity).
This exercise is made solely for lamniform sharks because tooth position and unit identifications in carcharhiniforms are 1) mostly lacking and 2) not easily differentiable in the fossil record.

A total of **552** lamniform tooth specimens lack positional information.
Because of inconsistencies in shark dental terminology we have followed the recommendations of Siverson 1999, who categorized ‘symphyseal’ teeth as ‘parasymphyseal’, and both ‘anterolateral’ and ‘lateral’ teeth as ‘lateroposterior’. However, we have retained an additional ‘posterior’ tooth morphotype grouping (sensu Cappetta 2012).
The following categories were not considered: **commissural (N=5) and intermediate (N=22)**.
```{r echo=TRUE,results='hide'}
# New column.
Lamn.Df$St.Position <- NULL
Lamn.Df$St.Position <- Lamn.Df$Relative.Position
Lamn.Df$St.Position <- factor(Lamn.Df$St.Position)

# Tooth positions.
Lamn.positions <- c("parasymphyseal","anterior","lateroposterior","posterior")
Lamn.monognathic <- Lamn.Df[Lamn.Df$St.Position %in% Lamn.positions, ]
Lamn.monognathic$St.Position <- factor(Lamn.monognathic$St.Position)
Lamn.monognathic$Age.2 <- factor(Lamn.monognathic$Age.2)

# Dental jaw units.
Lamn.units <- c("upper","lower")
Lamn.dignathic <- Lamn.Df[Lamn.Df$Dental.Unit %in% Lamn.units, ]
Lamn.dignathic$Dental.Unit <- factor(Lamn.dignathic$Dental.Unit)
Lamn.dignathic$Age.2 <- factor(Lamn.dignathic$Age.2)

# Combine positions and units.
combined.factors <- Lamn.dignathic[Lamn.dignathic$St.Position %in% Lamn.positions, ]

# Monognathic data frame (sample of 1345 specimens).
monognathic.Df <- geomorph.data.frame(gpa.coords = GPA$coords[,,rownames(Lamn.monognathic)],
                                      position = factor(Lamn.monognathic$St.Position),
                                      time = factor(Lamn.monognathic$Age.2))
# Dignathic data frame (sample of 1196 specimens).
dignathic.Df <- geomorph.data.frame(gpa.coords = GPA$coords[,,rownames(Lamn.dignathic)],
                                    unit = factor(Lamn.dignathic$Dental.Unit),
                                    time = factor(Lamn.dignathic$Age.2))
# Combined data frame (sample of 907 specimens).
combined.Df <- geomorph.data.frame(gpa.coords = GPA$coords[,,rownames(combined.factors)],
                                   unit = factor(combined.factors$Dental.Unit),
                                   position = factor(combined.factors$St.Position),
                                   time = factor(combined.factors$Age.2))

# Model 1: tooth position (monognathic) as a nominal factor.
T.1.Disp <- morphol.disparity(gpa.coords~position+time,groups = ~time,iter = 999,print.progress = T,data = monognathic.Df)
# Model 2: tooth position (dignathic) as a nominal factor.
T.2.Disp <- morphol.disparity(gpa.coords~unit+time,groups = ~time,iter = 999,print.progress = T,data = dignathic.Df)
# Model 3: tooth position and dental-unit (monognathic and dignathic) as nominal factors.
T.3.Disp <- morphol.disparity(gpa.coords~unit*position+time,groups = ~time,iter = 999,print.progress = T,data = combined.Df)
```

### Model Comparisons, in terms of the log-likelihood or covariance trace
```{r echo=TRUE,eval=F,results="hide"}
# Determine the best fit of disparity models.
fit.lm.1 <- lm.rrpp(f1 = two.d.array(gpa.coords) ~ time,iter = 999,RRPP = T,
                    data = combined.Df,print.progress = F)
fit.lm.2 <- lm.rrpp(f1 = two.d.array(gpa.coords) ~ time+position,iter = 999,RRPP = T,
                    data = combined.Df,print.progress = F)
fit.lm.3 <- lm.rrpp(f1 = two.d.array(gpa.coords) ~ time+unit,iter = 999,RRPP = T,
                    data = combined.Df,print.progress = F)
fit.lm.4 <- lm.rrpp(f1 = two.d.array(gpa.coords) ~ time+position+unit,iter = 999,RRPP = T,
                    data = combined.Df,print.progress = F)
# ANOVA
anova(fit.lm.1,fit.lm.2,fit.lm.3,fit.lm.4,effect.type = "F")
# Model Comparisons
modComp1 <- model.comparison(fit.lm.1,fit.lm.2,fit.lm.3,fit.lm.4,type = "logLik")
# Summary
summary(modComp1)
```

### Fourth-order sensitivity analysis
- Filter-feeding lamniform species (=35 observations) exist in our dataset. 
- We removed these observations and tested it's overall effect on total lamniform disparity through time.
```{r message=FALSE,warning=FALSE,cache=FALSE,eval=FALSE}
zooplan.tx <- which(Lamn.Df$Genus == "Megachasma" | Lamn.Df$Genus == "Cetorhinus"| Lamn.Df$Genus == "Cetorhinus ")
ex.frame <- Lamn.Df[-zooplan.tx, ]
# Geomorph data frame.
gd.ex <- geomorph.data.frame(gpa.coords = GPA$coords[,,rownames(ex.frame)],
                             age = factor(ex.frame$Age.2))
# Disparity.
non.filter.lam <- morphol.disparity(gpa.coords~age,groups = ~age,iter = 999,print.progress = T,data = gd.ex)

# Compare results.
plot(PV.Ple$Procrustes.var[c(3,8,5,16,17,7,2,12,14,4,1,6,15,9,18,10,11,13)],type = "o",
     ylim = c(0.01,0.1),ylab = "Procrustes variance",lwd = 1.5)
lines(non.filter.lam$Procrustes.var[c(3,8,5,16,17,7,2,12,14,4,1,6,15,9,18,10,11,13)],
      type = "o", col = "blue", lty = 2)
legend(x = 2,y = 0.02, 
       legend = c("Macrophages and microphagous lamniforms","Exclusion of microphagous lamniforms"),
       col = c("black","blue"),lty = c(19,19,19),lwd = 1.5,cex = .5)
```

### Fifth-order sensitivity analysis
Re-compute disparity of extant samples while accounting for heterodonty and families.
Sample size differ substantially between modern lamniform and carcharhiniform sharks.
Monognathic heterodont positions are largely missing for both groups. Conversely, the affinity of teeth from either upper or lower jaw units is well documented.
```{r message=FALSE,warning=FALSE,cache=FALSE,eval=FALSE}
Extant.Df <- droplevels(Data[which(Data$Age.2 == "Recent"), ])
Hetero.Df <- Extant.Df[Extant.Df$Dental.Unit %in% c("upper","lower"), ]
Hetero.Df$Dental.Unit <- factor(Hetero.Df$Dental.Unit)
Hetero.Df$Family <- factor(Hetero.Df$Family)

# Geomorph data frame.
Rig <- geomorph.data.frame(coords = GPA$coords[,,rownames(Hetero.Df)],
                           grp = factor(Hetero.Df$Order),
                           fam = factor(Hetero.Df$Family),
                           jaw = factor(Hetero.Df$Dental.Unit),
                           geography = factor(Hetero.Df$Country))
# Disparity.
a0 <- morphol.disparity(f1 = coords ~ 1,groups = ~ grp,iter = 999,data = Rig)
a1 <- morphol.disparity(f1 = coords ~ grp+jaw,groups = ~ grp,iter = 999,data = Rig)
a2 <- morphol.disparity(f1 = coords ~ grp+fam,groups = ~ grp,iter = 999,data = Rig)
a3 <- morphol.disparity(f1 = coords ~ grp+fam+jaw,groups = ~ grp,iter = 999,data = Rig)
a4 <- morphol.disparity(f1 = coords ~ grp+fam+jaw+geography,groups = ~ grp,iter = 999,data = Rig)
```

### Comparions of Holocene disparity
```{r message=FALSE,warning=FALSE,cache=FALSE,eval=FALSE}
comp.disp <- data.frame(Extant = c(0.06970290,0.05880965),
                        Dignathic.Effect = c(0.06742137,0.05875735),
                        Family = c(0.05182647,0.05270509),
                        Family.Jaw = c(0.05141668,0.05074020),
                        Species.Averages = c(0.03214946,0.03155608),
                        grp = c("Lamniformes","Carcharhiniformes"))

comp.disp <- melt(comp.disp)
comp.disp$grp <- relevel(comp.disp$grp,ref = "Lamniformes")
comp.disp %>%
  ggplot() + 
  geom_bar(mapping = aes(x = grp,y = value, fill = grp),
                      stat="identity") +
  scale_fill_manual(values = c("#E84646","#377EB8")) +
  theme(aspect.ratio = 1.5) +
  facet_grid(.~ variable,scales = "free")
```

### Visualize disparity profiles from sensitivity analysis (2)
```{r}
disp.hetero <- read.xlsx(file = "Metadata/Disparity vs. Heterodonty.xlsx",sheetIndex = 1)
disp.hetero$Age <- factor(disp.hetero$Age,levels = bins)
disp.hetero$Models <- factor(disp.hetero$Models,levels = c("Full","Monognathic","Dignathic","Combined"))
# Stacked bar plot with multiple groups
ggplot(data=disp.hetero, aes(x=Age, y=Disparity, group=Models,color=Models,shape=Models)) +
  geom_line(size = 2) + ylab("Procrustes Variance") +
  scale_fill_manual(values = c("#E16A86","#909800","#00AD9A","#9183E6")) +
  labs(title = "Comparative model evaluation") +
  theme(axis.title = element_text(size = 7),
        strip.text.x = element_text(size = 7, face = "bold"),
        axis.text.y = element_text(angle = 90, size = 7),
        axis.text.x = element_text(size = 7,angle = 45,vjust = .7),
        panel.spacing = unit(0, "lines"),
        aspect.ratio = 1)
```

### Climate and disparity data
- Select the relevant time-interval from the climate dataset.
- Temperature and sea-level data frames are used to visualize abiotic changes through time.
- The *disp.res* data frame summarizes environmental proxy by time and clade disparity, and is used for subsequent GLS-analysis.
- The link between shark tooth morphology and climate is not known; but since it is the phenotype that interacts with the environment, we predict that morphological trajectories will be influenced by changes in the environment.
```{r}
# Import data frames.
temperature <- read.xlsx(file = "Metadata/Condamine et al. Climate Data.xlsx",sheetIndex = 2,header = T)
# Campanian-Recent interval.
temperature <- temperature[temperature$Age <= 83.6 & temperature$Age >= 0, ]
seaLevel <- read.xlsx(file = "Metadata/Condamine et al. Climate Data.xlsx",sheetIndex = 3,header = T)
# Campanian-Recent interval.
seaLevel <- seaLevel[seaLevel$Age <= 83.6 & seaLevel$Age >= 0, ]
# Disparity results with PI-intervals.
disp.res <- read.xlsx(file = "Metadata/GLS Data.xlsx",sheetIndex = 1)
# Cenozoic Global Reference benthic foraminifer carbon and oxygen Isotope Dataset (CENOGRID).
west.Df <- read.csv2(file = "Metadata/Westerhold et al. 2020.csv",header = T)
# Miller et al. 2020
Miller.2020 <- read.xlsx(file = "Metadata/Miller et al. 2020.xlsx",sheetIndex = 1)
Miller.2020$Age <- as.numeric(Miller.2020$Age)
Miller.2020$Oxygen <- as.numeric(Miller.2020$Oxygen)
Miller.2020$Sea.Level <- as.numeric(Miller.2020$Sea.Level)
Miller.2020$Age.Ma <- NULL
Miller.2020$Age.Ma <- Miller.2020$Age/1000
# Miller et al. 2005
Miller.2005 <- read.csv(file = "Metadata/Miller et al. 2005.txt",header = T,sep = "\t",dec = ".")
Miller.2005 <- Miller.2005[,-1]
Miller.2005 <- Miller.2005[Miller.2005$Age_Miller <= 83.6 & Miller.2005$Age_Miller >= 0, ]
# Cramer et al. 2009 (benthic foraminiferal d18O)
Cramer.2009 <- read.csv(file = "Metadata/Cramer et al. 2009.txt",header = T,sep = "\t")
Cramer.2009 <- Cramer.2009[Cramer.2009$ï..Age..GTS2004. <= 83.6 & Cramer.2009$ï..Age..GTS2004. >= 67.101133, ]
Cramer.2009 <- Cramer.2009[rowSums(is.na(Cramer.2009)) != ncol(Cramer.2009), ] 
```

```{r}
layout(matrix(1:3,nrow = 3,ncol = 2),respect = T)
# Plot Mesozoic and Cenozoic temperature from Prokoph et al 2008 & Zachos et al. 2008
plot(x = temperature$Age,y = temperature$Temperature,type = "l",
     xlim = rev(range(temperature$Age,na.rm = T)),
     ylim = range(temperature$Temperature),
     axes = T, xlab = "", ylab = "Deep-sea temperature (\u00B0C)", col = "#CC6666",lwd = 2)
abline(v = c(72.0,66.0,56.0),lwd = 2)
# Plot CENOGRID	benthic	foraminifer	oxygen isotope splice from Westerhold et al. (2020)
plot(west.Df$age_tuned, west.Df$benthic.d18O.VPDB, type = "l",
     xlim = rev(range(west.Df$age_tuned,na.rm = T)),
     ylim = rev(range(west.Df$benthic.d18O.VPDB,na.rm = T)),
     axes = T, xlab = "", ylab = expression(paste("benthic ",delta^{18}, "O (‰)")), col = "#9999CC", lwd =2)
abline(v = c(67.10,66.0,56.0),lwd = 2)
lines(x = Cramer.2009$ï..Age..GTS2004., y = Cramer.2009$d18O,col = "black",type = "l",lwd = 2)
text(x = c(50,75),y = c(2,2),labels = c("CENOGRID","Cramer et al. 2009"),cex = .5)
# Cramer et al. 2009 (deep ocean temperature variations)
plot(Cramer.2009$ï..Age..GTS2004.,Cramer.2009$d18O,type = "l",
     xlim = rev(range(Cramer.2009$ï..Age..GTS2004.,na.rm = T)),
     ylim = rev(range(Cramer.2009$d18O,na.rm = T)),
     ylab = expression(paste("benthic ",delta^{18}, "O (‰)")), col = "#9999CC", lwd =2)
abline(v = c(56.0,66.0,72.1,77))
dev.off()
```

```{r}
# Plot sea-level from Miller et al. 2020
# NB:
# Sea level: computed using delta 18O seawater.
plot(Miller.2020$Age/1000,Miller.2020$Sea.Level,
     type = "l", xlim = rev(range(Miller.2020$Age/1000)))
abline(v = c(66.0,56.0));title(main = "Miller et al. 2020")

# The Phanerozoic Record of Global: Miller et al. 2005
# Raw.
par(mfrow = c(2,1))
plot(Miller.2005$Age_Miller,Miller.2005$SL_Miller,
     type = "l",xlim = rev(range(Miller.2005$Age_Miller)))
abline(v = c(66.0,56.0));title(main = "Miller et al. 2005")
# From Condamine et al. 2019
plot(seaLevel$Age,seaLevel$SeaLevel,
     type = "l",xlim = rev(range(seaLevel$Age)))
abline(v = c(66.0,56.0));title(main = "Condamine et al. 2019")
dev.off()
```

```{r}
par(mar = c(5,2.5,1,5), oma = c(0,0,0,2))
# Astronomically tuned deep-sea benthic foraminifer oxygen (delta-18O) isotope data.
plot(west.Df$age_tuned, west.Df$benthic.d18O.VPDB, type = "l", 
     xlim = rev(range(west.Df$age_tuned,na.rm = T)), axes = FALSE, 
     xlab = "", ylab = "n", col = "black")
axis(side = 4, ylab = "", col = "black")
lines(x = Cramer.2009$ï..Age..GTS2004., y = Cramer.2009$d18O,col = "maroon",type = "l",lwd = 2)
par(new = TRUE)
# Sea-level: benthic isotope stack previously compiled by Miller et al. (2020)
plot(Miller.2020$Age/1000, Miller.2020$Sea.Level, type = "l", 
     xlim = rev(range(west.Df$age_tuned,na.rm = T)),
     axes = F, xlab = "", ylab = "n", col = "#9999CC")
axis(side = 4, ylab = "", col = "#9999CC", line = 2.5)
axis(side = 1)
box(lty = 1, col = 'black')
legend("bottomright", legend = c("Sea Level (m)","Oxygen Isotope (dO18)"), 
       col = c("black", "blue"), lty = 1, cex = 0.65)
abline(v = c(66,56.0,53.4,49.2,34,17,14,3.3),lwd = 2)
dev.off()
```

### Visualize disparity and environmental variables
```{r warning=FALSE}
# Temperature profile along with a first degree (linear) polynomial.
# TEX86-based	SST.
p1 <- ggplot(data = temperature, aes(x = Age, y = Temperature)) +
  scale_x_reverse() +
  geom_point(data = disp.res, aes(x = Midpoint,y = d18O),pch = 21, col = "black", size = 2) + # add weighted averages.
  geom_line(aes(y = Temperature),colour  = 'deeppink2', size = 1, alpha = .5) +
  geom_smooth(method = 'loess',se = TRUE, col = "black",span = 0.3,formula = y ~ poly(x, degree = 1, raw = TRUE)) +
  geom_vline(xintercept = c(83.6,72.1,66.0,61.6,59.2,56.0,
                            47.8,41.2,37.8,33.9,27.82,23.03,
                            20.44,15.97,13.82,11.63,7.246,
                            5.333,3.600,2.58,0.0117,0), color = "orange") +
  theme_classic()

# Temperature (CENOGRID Data)
# NB: For data with > 1,000 observations: the default model is a GAM (a general additive model)
p.0 <- ggplot(data = na.omit(west.Df), aes(x = age_tuned, y = benthic.d18O.VPDB)) +
  coord_cartesian(xlim = rev(c(0.00585,77.85000))) +
  scale_x_reverse() +
  scale_y_reverse() +
  geom_line(aes(y = benthic.d18O.VPDB),colour  = '#9999CC', size = 1) +
  geom_point(data = disp.res, aes(x = Midpoint,y = d18O.2020),pch = 21, col = "black", size = 2) +
  # Local weighted regression with a 10% smoothing span.
  geom_smooth(method = 'loess',se = TRUE, col = "black",span = 0.3,formula = y ~ poly(x, degree = 1, raw = TRUE)) +
  # stat_smooth(method = "loess",span = 0.10,col = "black",se = T,fullrange = T) +
  geom_vline(xintercept = c(67.10,56.0), color = "orange") +
  theme_classic()

# Sea-level curve (Miller et al. 2020)
p2 <- ggplot(data = Miller.2020, aes(x = Age.Ma, y = Sea.Level)) +
  coord_cartesian(xlim = rev(c(0.00585,77.85000))) +
  scale_x_reverse() +
  geom_line(aes(y = Sea.Level),colour  ="#C54E6D", size = 1) +
  geom_point(data = disp.res, aes(x = Midpoint,y = SL.2020),pch = 21, col = "black", size = 2) +
  geom_vline(xintercept = c(66.61089,56.0,53.4,49.2,17,14), color = "black") +
  theme_classic()

p.S <- ggplot(data = seaLevel, aes(x = Age, y = SeaLevel)) +
  scale_x_reverse() +
  geom_line(aes(y = SeaLevel),colour  ="#C54E6D", size = 1) +
  geom_point(data = disp.res, aes(x = Midpoint,y = SL.2020),pch = 21, col = "black", size = 2) +
  geom_vline(xintercept = c(66.61089,56.0,53.4,49.2,17,14), color = "black") +
  theme_classic()

# Tooth disparity.
pd <- position_dodge(width = 2.5)
# disp.res$Clade <- relevel(disp.res$Clade ,"Lamniformes")
# Plot
p3 <- ggplot(disp.res, aes(x = Midpoint, y = Disparity, colour = Clade)) +
  # coord_cartesian(xlim = rev(c(0.00585,77.85000))) +
  scale_x_reverse() +
  geom_vline(xintercept = c(83.6,72.1,66.0,61.6,59.2,56.0,
                            47.8,41.2,37.8,33.9,27.82,23.03,
                            20.44,15.97,13.82,11.63,7.246,
                            5.333,3.600,2.58,0.0117,0), color = "orange") +
  theme_grey() +
  theme(axis.title.x = element_text(face = "bold"),
        axis.title.y = element_text(face = "bold"),legend.position = "none") +
  geom_errorbar(aes(ymin = LowerBoot.PI ,ymax = UpperBoot.PI), width = 1.5, position = pd, show.legend = FALSE,size = .5) +
  geom_line(position = pd) +
  geom_point(position = pd, size = 3, shape = 21, fill = "white") +
  scale_colour_manual(values = c("#377EB8","#E84646")) +
  ylab("Procrustes Variance") + xlab("")

# Combine
# c.plots <- list(p.0, p2, p3)
# plot_grid(plotlist=c.plots, ncol=1, align='v',rel_heights = c(30,30,60))

# Add a shaped rectangle.
# p3 <- p3 + annotate("rect", xmin = c(23.03), xmax = c(5.333),ymin = c(0.02), ymax = c(0.08), alpha = .1, fill = "blue")

# Combine graphs.
# grid.draw(rbind(ggplotGrob(p2), ggplotGrob(p.0), size = "last"))
# grid.draw(rbind(ggplotGrob(p2), ggplotGrob(p3), size = "last"))
# grid.draw(rbind(ggplotGrob(p.0), ggplotGrob(p2), ggplotGrob(p3),size = "last"))
```

### Effect of variable time-bin duration
- Time duration (Ma) has no effect on the estimated disparity of Lamniformes but appear to have a minor effect in Carcharhiniformes.
- *Does disparity increase with longer time duration?*
```{r}
# New column.
disp.res$Duration <- NULL
disp.res$Duration <- disp.res$Start - disp.res$End
# Fit model.
summary(lm(disp.res$Disparity[1:18]~disp.res$Duration[1:18],
           weights = Sample[1:18],data = disp.res))
summary(lm(disp.res$Disparity[19:36]~disp.res$Duration[19:36],
            weights = Sample[19:36],data = disp.res))
# Plot
ggplot(disp.res,aes(x = Duration, y = Disparity)) +
  geom_point() + xlab("Duration (Ma)") + ylab("Procrustes Variance") +
  stat_smooth(method = "lm",formula = y ~ x) + facet_wrap(.~Clade,scales = "free") +
  theme(axis.title = element_text(size = 7),
        strip.text.x = element_text(size = 7, face = "bold"),
        axis.text.y = element_text(angle = 90, size = 7),
        axis.text.x = element_text(size = 7),
        panel.spacing = unit(0, "lines"),
        aspect.ratio = 1)
```

### Rarefied disparity
```{r message=FALSE,warning=FALSE,cache=FALSE,eval=FALSE}
# 1. Lamniformes.
lamn.rare.res <- read.xlsx(file = "Table Results/Lamn.Rarefaction.xlsx",sheetIndex = 1)
lamn.rare.res$Age <- factor(lamn.rare.res$Age,levels=c("Campanian","Maastrichtian","Danian/Selandian","Thanetian",
                                                       "Ypresian","Lutetian","Bartonian","Priabonian","Rupelian","Chattian",
                                                       "Aquitanian/Burdigalian","Langhian","Serravallian/Tortonian",
                                                       "Messinian","Zanclean","Piacenzian","Recent"))
# 2. Carcharhiniformes.
carch.rare.res <- read.xlsx(file = "Table Results/Carch.Rarefaction.xlsx",sheetIndex = 1)
carch.rare.res$Age <- factor(carch.rare.res$Age,levels=c("Campanian","Maastrichtian","Danian/Selandian",
                                                         "Thanetian","Ypresian","Lutetian","Bartonian","Priabonian",
                                                         "Rupelian","Chattian","Aquitanian/Burdigalian",
                                                         "Langhian","Serravallian/Tortonian","Messinian",
                                                         "Zanclean","Pleistocene","Piacenzian","Recent"))

# Sub-sampling-level in Lamniformes is the Piacenzian (N=12).
pos <- position_dodge2(width = 0.2)
ggplot(lamn.rare.res,mapping = aes(x = N, y = rare.disp)) +
  geom_errorbar(aes(ymin = lower.CI ,ymax = upper.CI), 
                position = pos, col = "#AA1936") +
  geom_point(shape = 21, colour = "#AA1936", fill = "white",position = pos) +
  facet_wrap(~Age,scales = "free",ncol = 4) + ylab("Disparity") +
  labs(title = "Lamniformes",subtitle = "Temporal sample-rarefaction curves") +
  theme(strip.text.x = element_text(size = 7, face = "bold"),
        axis.title = element_text(color = "#666666", face = "bold", size = 7),
        axis.text.y = element_text(angle = 90, size = 7,),
        axis.text.x = element_text(size = 7),
        title = element_text(size = 7)) +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1))

# Sub-sampling-level in Carcharhiniformes is the Bartonian and Piacenzian (N=8).
ggplot(carch.rare.res,mapping = aes(x = N, y = rare.disp)) +
  geom_errorbar(aes(ymin = lower.CI ,ymax = upper.CI), 
                position = pos, col = "darkblue") +
  geom_point(shape = 21, colour = "darkblue", fill = "white",position = pos) +
  facet_wrap(~Age,scales = "free",ncol = 4) + ylab("Disparity") +
  labs(title = "Carcharhiniformes",subtitle = "Temporal sample-rarefaction curves") +
  theme(strip.text.x = element_text(size = 7, face = "bold"),
        axis.title = element_text(color = "#666666", face = "bold", size = 7),
        axis.text.y = element_text(angle = 90, size = 7),
        axis.text.x = element_text(size = 7),
        title = element_text(size = 7)) +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1))
```

### Alternative rarefaction profiles
```{r, warning=FALSE}
rarefaction.df <- read.xlsx(file = "Table Results/Ts.Rarefaction.xlsx",sheetIndex = 1)
# rarefaction.df$Clade <- relevel(rarefaction.df$Clade ,"Lamniformes")

# Plot.
ggplot(rarefaction.df, aes(x = Midpoint, y = Rarefied.Disparity, colour = Clade)) +
  scale_x_reverse() +
  theme_grey() +
  theme(axis.title.x = element_text(face = "bold"),
        axis.title.y = element_text(face = "bold"),legend.position = "none") +
  geom_errorbar(aes(ymin = LowerBoot.PI ,ymax = UpperBoot.PI), width = 1.5, position = pd,
                show.legend = FALSE,size = .5,linetype = "dashed") +
  geom_line(position = pd) +
  geom_point(position = pd, size = 3, shape = 21, fill = "white") +
  scale_colour_manual(values = c("#E84646","#377EB8")) +
  ylab("Procrustes Variance") + xlab("") + labs(title = "Rarefied dental disparity") +
  theme(panel.spacing = unit(0, "lines")) +
  theme(aspect.ratio = 1/2)
```

### Generalized least-squares models (GLS)
We implemented customized *gls models* functions based on the *nlme* v.3.1-148 package (Pinheiro et al. 2020) to compare five generalized least squares (GLS) models to describe patterns of shark tooth disparity through time.

1.	*‘Null model’*. This provided a single parameter intercept-only model whereby shark tooth disparity was described as stochastic (= Brownian motion).
2.	*‘Sea-level model’*. This provided a two-parameter slope+intercept model that described a relationship between changing sea-level and shark tooth disparity.
3.	*‘delta 18O model’*. This provided a second two-parameter slope+intercept model that described a relationship between changing delta 18O paleotemperature values and shark tooth disparity.
4.	*‘Sea-level+delta 18O model’*. This provided a three-parameter slope+slope+intercept model that described a multivariate relationship between changing sea-level, delta 18O paleotemperature values, and shark tooth disparity.
5.	*‘Sea-level:delta 18O model’*. This provided a four-parameter slope+slope+interaction+intercept model that described a multivariate relationship between changing sea-level, delta 18O paleotemperature values, and the interaction between changing sea-level and delta 18O paleotemperature values, together with shark tooth disparity.

GLS model fitting uses maximum likelihood with alternate implementation of a first-order autoregressive model (AR1). Because values along a time-series tend to be serially correlated, and may thus result in autocorrelation (= biased residuals under a standard linear model), as well as increased type I errors, we additionally screened our data with an autocorrelation function (ACF) to calculate the amount of serial correlation between lagged residuals (Pinheir & Bates, 2000). These results were visually inspected via plots of 12 standardized lags mapped against their corresponding autocorrelations. Significant correlation was tested through a Ljung-Box Test (Ljung & Box, 1978) and the models were subsequently adjusted to accommodate. Our final interpretations are, therefore, based only on those models that integrated a time-calibrated AR1 correlation structure within their GLS framework. The GLS likelihood criteria included the Akaike information criterion (AIC), the Akaike information criterion for small samples (AICc), the Akaike weights, and likelihood ratio tests between similarly supported models. Values for the customized functions also incorporated a list of results from the autocorrelation function, together with the model parameters, and a summary table of the likelihood statistics modified from anova.gls.
```{r}
# Time-series data frame.
GLS.Df <- read.xlsx(file = "Metadata/GLS Data.xlsx",sheetIndex = 1)
# Get time-bin duration.
GLS.Df$Duration <- NULL
GLS.Df$Duration <- GLS.Df$Start - GLS.Df$End
# Lamniformes gls data.
df.lamn <- GLS.Df[1:18,]
# Carcharhiniformes gls data.
df.carch <- GLS.Df[19:36,]
```

```{r}
# Two-way interaction plot: this plots mean disparity values.
with(GLS.Df, interaction.plot(Midpoint,Clade,Disparity,lty = 2, font.lab = 2,col = c("grey","black"),
                              fixed = T, leg.bty = "o",
                              type = "l",main = "Two-way Interaction Plot"))
dev.off()
```

```{r}
# 1.Lamniformes time-series.
# Preliminary OLS regression and autocorrelation function.
# It makes no difference when including "time" as a covariate.
l.mod <- lm(formula = Disparity~SL+d18O,data = df.lamn)
# Diagnostics.
par(mfrow = c(3,2))
plot(l.mod, which = 1:6)
dev.off()
```

```{r}
# Plot residuals against time: Supplementary Figure 34A.
plot(residuals(l.mod) ~ df.lamn$Midpoint,xlab = "Midpoint",ylab = "Normalized residuals",font.lab = 2, main = "Lamniformes", cex = .7)
abline(h = 0, lty = 2); lines(residuals(l.mod) ~ df.lamn$Midpoint, lwd = 1)
dev.off()
# NB: no visible pattern and/or trend.
```

```{r}
# Complete ACF plot: Supplementary Figure 34B.
acf(residuals(l.mod),main = "Complete autocorrelogram", sub = "Lamniformes",font.lab = 2)
dev.off()
# Previous values of a time series are called lags. 
```

```{r}
# PACF: Pre-GLS autocorrelation.
pacf(residuals(l.mod),lag.max = 12,cex = .7, cex.lab = .7, cex.axis = .7,main = "Lamniformes")
dev.off()
# Box–Pierce test: include results in graph.
Box.test(residuals(l.mod), lag = 12,type = "Box-Pierce") #  can't reject the null-hypothesis
Box.test(residuals(l.mod), lag = 12, type = "Ljung-Box") # reject the null-hypothesis
```

```{r}
# Visualize ‘auto-dependence’ 
lag.plot(residuals(l.mod),2,diag.col = "maroon", do.lines = FALSE)
dev.off()
# Main result: No autocorrelation is detected in Lamniformes time-series.
```

```{r}
# 2. Carcharhiniformes pre-GLS-models.
c.mod <- lm(formula = Disparity~d18O+SL,data = df.carch)
plot(residuals(c.mod) ~ df.carch$Midpoint, ylab = "Normalized residuals",main = "Carcharhiniformes", cex = .7)
abline(h = 0, lty = 2);lines(residuals(c.mod) ~ df.carch$Midpoint)
dev.off()
```

```{r}
# ACF plot: autocorrelogram.
acf(residuals(c.mod),main = "Complete autocorrelogram",sub = "Carchariniformes", font.lab = 2)
dev.off()
```

```{r}
# PACF.
pacf(residuals(c.mod),lag.max = 12,cex = .7, cex.lab = .7, cex.axis = .7,main = "Carcharhiniformes")
dev.off()
```

```{r}
# Box–Pierce test.
Box.test(residuals(c.mod), lag = 12,type = "Box-Pierce",fitdf = 1) #  can't reject the null-hypothesis
Box.test(residuals(c.mod), lag = 12, type = "Ljung-Box",fitdf = 1) #  can't reject the null-hypothesis
```

### GLS model fitting
- The following tests were conducted under the assumption that shark dental disparity is a measure of diversity, akin to taxonomic richness, and a proxy for ecology, both of which are likely affected by changes in sea-level (a proxy for habitat availability) and ocean temperature. For lamniforms and carcharhiniforms alike we predict that a drop in sea level will decrease dental disparity.
- GLS model fitting using the whole time-series (Campanian-Recent) and average climate data as compiled by Condamine et al. (2019).
If *phi* is negative, then the ACF is not monotonous in lags, but changes from positive to negative.
**NB:** A positive coefficient indicates that as the value of the independent variable increases, the mean of the dependent variable also tends to increase. A negative coefficient suggests that as the independent variable increases, the dependent variable tends to decrease.
```{r cache=FALSE,eval=FALSE}
# Lamniformes
w.gls.lamn <- data.frame(time = df.lamn$Midpoint,N = df.lamn$Disparity,
                         sea.level = df.lamn$SL,
                         delta.O18 = df.lamn$d18O)
# Carcharhiniformes
w.gls.carch <- data.frame(time = df.carch$Midpoint,N = df.carch$Disparity,
                          sea.level = df.carch$SL,
                          delta.O18 = df.carch$d18O)
# Models.
w.L.tot.noAR <- gls.model.full(w.gls.lamn, acf.plot = TRUE) # No autocorrelation structure.
w.L.tot.AR1 <- gls.model.full(w.gls.lamn, corStruct = TRUE, acf.plot = TRUE,p.adjust.type = "fdr") # With autocorrelation structure.
dev.off()
# Approximate confidence intervals
nlme:::intervals.gls(w.L.tot.noAR$gls.models$`Sea-Level`,which = "var-cov")
nlme:::intervals.gls(w.L.tot.AR1$gls.models$`Sea-Level`,which = "var-cov")
# Ljung-Box Test
Box.test(w.L.tot.AR1$gls.models$Null$residuals,type = "Ljung-Box")
Box.test(w.L.tot.AR1$gls.models$`Sea-Level`$residuals,type = "Ljung-Box")
Box.test(w.L.tot.AR1$gls.models$Temperature$residuals, type = "Ljung-Box")
Box.test(w.L.tot.AR1$gls.models$`SL+Temp`$residuals, type = "Ljung-Box")
Box.test(w.L.tot.AR1$gls.models$`SL*Temp`$residuals, type = "Ljung-Box")
# Models.
w.C.tot.noAR <- gls.model.full(w.gls.carch, acf.plot = TRUE) # No autocorrelation structure
w.C.tot.AR1 <- gls.model.full(w.gls.carch, corStruct = TRUE, acf.plot = TRUE,p.adjust.type = "fdr") # With autocorrelation structure
dev.off()
# Approximate confidence intervals
nlme:::intervals.gls(w.C.tot.noAR$gls.models$`Sea-Level`,which = "var-cov")
nlme:::intervals.gls(w.C.tot.AR1$gls.models$`Sea-Level`,which = "var-cov")
```

```{r cache=FALSE,eval=FALSE}
# Table results
kable(w.L.tot.noAR$Overall.Results, format = "pandoc", digits = c(0, 0, 0, 0, 0, 0, 4, 4, 4, 0, 4),
      caption = "Entire time series for lamniform sharks, with an AR0 model")
kable(w.L.tot.AR1$Overall.Results, format = "pandoc", digits = c(0, 0, 0, 0, 0, 0, 4, 4, 4, 0, 4),
      caption = "Entire time series for lamniform sharks, with an AR1 model")
kable(w.C.tot.noAR$Overall.Results, format = "pandoc", digits = c(0, 0, 0, 0, 0, 0, 4, 4, 4, 0, 4),
      caption = " Entire time series for carcharhiniform sharks, with an AR0 model")
kable(w.C.tot.AR1$Overall.Results, format = "pandoc", digits = c(0, 0, 0, 0, 0, 0, 4, 4, 4, 0, 4),
      caption = "Entire time series for carcharhiniform sharks, with an AR1 model")
```

```{r cache=FALSE,eval=FALSE}
mods <- c("Null", "Sea-level", "Delta.18O", "Sea-level+Delta.18O", "Sea-level*Delta.18O")
lamn.tot.AW <- w.L.tot.AR1$Overall.Results$AICc.W; names(lamn.tot.AW) <- w.L.tot.AR1$Overall.Results$Model.Factor
carch.tot.AW <- w.C.tot.AR1$Overall.Results$AICc.W; names(carch.tot.AW) <- w.C.tot.AR1$Overall.Results$Model.Factor
# Plot corrected Akaike Information Criterion weights
AICc.res <- matrix(data = c(lamn.tot.AW[mods], carch.tot.AW[mods]), ncol = 2,
                            dimnames = list(mods, c("Lamniformes", "Carcharhiniformes")))
barplot(as.matrix(AICc.res),col=brewer.pal(n = 5, name = "RdBu"),ylab = "AICc.W",font.lab = 2)
```

### The entire Cenozoic
```{r cache=FALSE,eval=FALSE}
# Subset
ceno.lamn <- df.lamn[-c(1:2),]
ceno.carch <- df.carch[-c(1:2),]
# GLS Frames
r.gls.lamn <- data.frame(time = ceno.lamn$Midpoint,N = ceno.lamn$Disparity,
                         sea.level = ceno.lamn$SL,
                         delta.O18 = ceno.lamn$d18O)

r.gls.carch <- data.frame(time = ceno.carch$Midpoint,N = ceno.carch$Disparity,
                          sea.level = ceno.carch$SL,
                          delta.O18 = ceno.carch$d18O)
# Models.
r.L.tot.noAR <- gls.model.full(r.gls.lamn, acf.plot = TRUE) # No autocorrelation structure.
r.L.tot.AR1 <- gls.model.full(r.gls.lamn, corStruct = TRUE, acf.plot = TRUE,p.adjust.type = "fdr") # With autocorrelation structure.
dev.off()
# Models.
r.C.tot.noAR <- gls.model.full(r.gls.carch, acf.plot = TRUE) # No autocorrelation structure
r.C.tot.AR1 <- gls.model.full(r.gls.carch, corStruct = TRUE, acf.plot = TRUE,p.adjust.type = "fdr") # With autocorrelation structure
dev.off()
```

```{r cache=FALSE,eval=FALSE}
# Table results
kable(r.L.tot.noAR$Overall.Results, format = "pandoc", digits = c(0, 0, 0, 0, 0, 0, 4, 4, 4, 0, 4),
      caption = "Cenozoic series for lamniform sharks, with an AR0 model")
kable(r.L.tot.AR1$Overall.Results, format = "pandoc", digits = c(0, 0, 0, 0, 0, 0, 4, 4, 4, 0, 4),
      caption = "Cenozoic series for lamniform sharks, with an AR1 model")
kable(r.C.tot.noAR$Overall.Results, format = "pandoc", digits = c(0, 0, 0, 0, 0, 0, 4, 4, 4, 0, 4),
      caption = "Cenozoic series for carcharhiniform sharks, with an AR0 model")
kable(r.C.tot.AR1$Overall.Results, format = "pandoc", digits = c(0, 0, 0, 0, 0, 0, 4, 4, 4, 0, 4),
      caption = "Cenozoic series for carcharhiniform sharks, with an AR1 model")
```

### The entire Cenozoic using the CENOGRID and delta-18O benthic splice data
```{r cache=FALSE,eval=FALSE}
# GLS Frames
grid.gls.lamn <- data.frame(time = ceno.lamn$Midpoint,N = ceno.lamn$Disparity,
                            sea.level = ceno.lamn$SL.2020,
                            delta.O18 = ceno.lamn$d18O.2020)

grid.gls.carch <- data.frame(time = ceno.carch$Midpoint,N = ceno.carch$Disparity,
                             sea.level = ceno.carch$SL.2020,
                             delta.O18 = ceno.carch$d18O.2020)
# Models.
grid.L.tot.noAR <- gls.model.full(grid.gls.lamn, acf.plot = TRUE) # No autocorrelation structure.
grid.L.tot.AR1 <- gls.model.full(grid.gls.lamn, corStruct = TRUE, acf.plot = TRUE,p.adjust.type = "fdr") # With autocorrelation structure.
dev.off()
# Models.
grid.C.tot.noAR <- gls.model.full(grid.gls.carch, acf.plot = TRUE) # No autocorrelation structure
grid.C.tot.AR1 <- gls.model.full(grid.gls.carch, corStruct = TRUE, acf.plot = TRUE,p.adjust.type = "fdr") # With autocorrelation structure
dev.off()
```

```{r cache=FALSE,eval=FALSE}
# Lamniformes
kable(grid.L.tot.noAR$Overall.Results, format = "pandoc", digits = c(0, 0, 0, 0, 0, 0, 4, 4, 4, 0, 4),
      caption = "Cenozoic series for lamniform sharks, with an AR0 model")
kable(grid.L.tot.AR1$Overall.Results, format = "pandoc", digits = c(0, 0, 0, 0, 0, 0, 4, 4, 4, 0, 4),
      caption = "Cenozoic series for lamniform sharks, with an AR1 model")
# Carcharhiniformes
kable(grid.C.tot.noAR$Overall.Results, format = "pandoc", digits = c(0, 0, 0, 0, 0, 0, 4, 4, 4, 0, 4),
      caption = "Cenozoic series for carcharhiniform sharks, with an AR0 model")
kable(grid.C.tot.AR1$Overall.Results, format = "pandoc", digits = c(0, 0, 0, 0, 0, 0, 4, 4, 4, 0, 4),
      caption = "Cenozoic series for carcharhiniform sharks, with an AR1 model")
```

### Family-level morphospace analysis
- Exploration of family-level sub-morphospace(s) through time.
- All extant specimens identified as **Carcharias taurus** belong to the family Carchariidae and not Odontaspididae (see Stone & Shimada 2019).
- Patterns are shown using density distributions along with the mean and modal values are computed for each family.
```{r Lamniformes families,warning=FALSE}
# Clean-up family level entries.
# Omit Eoptolamnidae and Cretoxyrhinidae (N=1), and time-bins with sample < 2.
# Jaekelotodontidae could be synonymous with Odontaspididae.
Lamn.Fam.Df <- Lamn.Df[Lamn.Df$Family != "" & Lamn.Df$Family != "Family incertae sedis" & 
                       Lamn.Df$Family != "Eoptolamnidae" & Lamn.Df$Family != "Cretoxyrhinidae", ]
Lamn.Fam.Df$Family <- factor(Lamn.Fam.Df$Family)
Lamn.Fam.Df$Age.2 <- factor(Lamn.Fam.Df$Age.2)
levels(Lamn.Fam.Df$Family)[6] <- "Lamnidae" # Isuridae is synonymous with Lamnidae.
nlevels(Lamn.Fam.Df$Family) # 15 levels.
# Set up new data frame.
gg.lamn.fam <- data.frame(pc1 = Lamn.scores[rownames(Lamn.Fam.Df),1],
                          pc2 = Lamn.scores[rownames(Lamn.Fam.Df),2],
                          pc3 = Lamn.scores[rownames(Lamn.Fam.Df),3],
                          family = factor(Lamn.Fam.Df$Family),
                          age = factor(Lamn.Fam.Df$Age.2))

# Stratigraphically re-arrange time-bins.
gg.lamn.fam$age <- with(gg.lamn.fam,factor(gg.lamn.fam$age,levels = rev(bins)))
# Arrange families by extant and extinct units.
gg.lamn.fam$family <- factor(gg.lamn.fam$family,
                             levels = c("Anacoracidae","Archaeolamnidae",
                                        "Jaekelotodontidae",
                                        "Otodontidae","Pseudocoracidae",
                                        "Serratolamnidae","Xiphodolamiidae","Alopiidae",
                                        "Carchariidae","Cetorhinidae","Lamnidae",
                                        "Megachasmidae","Mitsukurinidae","Odontaspididae",
                                        "Pseudocarchariidae"))

# Compute family-level means at each time-interval.
# Replace pc1 to get summary values of other axes.
# na.rm set to TRUE, because of the presence of empty intervals.
mean.lamn.fam <- gg.lamn.fam %>% 
  group_by(family, age) %>%
  dplyr::summarise(grp.mean = mean(pc1),
                   grp.median = median(pc1),
                   grp.mode = estimate_mode(pc1),na.rm = T)

# Plot density distributions over time.
# Absolute x-values differ between groups, and reflect the maximum attained range over time.
# The same applies to the density which is determined by the number of observations given time (T[x]).
ggplot(gg.lamn.fam, aes(x = pc1,fill = family, colour = family)) +
  geom_histogram(bins = 30) +
  scale_y_continuous(labels = scales::number_format(accuracy = 1)) +
  scale_x_continuous(n.breaks = 4) +
  # geom_density(alpha=.5,lwd = .5) + 
  geom_vline(aes(xintercept = grp.mean), 
             linetype = "dashed", size = 0.4,
             color = "black", data = mean.lamn.fam) + geom_vline(aes(xintercept = grp.mode), 
             linetype = "solid", size = 0.4,
             color = "#EFC000FF", data = mean.lamn.fam) +
  scale_color_discrete_sequential(palette = "Red-Blue",nmax = 15) +
  scale_fill_discrete_sequential(palette = "Red-Blue",nmax = 15) +
  theme_grey() + xlab("Principal Component 1 (64.78%)") + ylab("Frequency") +
  theme(strip.text.x = element_text(size = 7, angle = 75, face = "bold"),
        strip.text.y = element_text(size = 7, face = "bold"),
        axis.title = element_text(color = "black", face = "bold", size = 7),
        strip.background = element_rect(colour = "black")) +
  theme(axis.text.x = element_text(size = 6),
        axis.text.y = element_text(size = 6),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        legend.position = "none") + facet_grid(age ~ family,scales = "free",space = "free_x")
```

```{r Carcharhiniformes families,warning=FALSE}
# Leptochariidae (N=2) and Pseudotriakidae (N=3) is removed because of low sample size.
# Subset.
Carch.Fam.Df <- Carch.Df[Carch.Df$Family != "" & Carch.Df$Family != "Family incertae sedis" &
                         Carch.Df$Family != "indet." & Carch.Df$Family != "Leptochariidae" &
                         Carch.Df$Family != "Pseudotriakidae", ]

Carch.Fam.Df$Family <- factor(Carch.Fam.Df$Family)
levels(Carch.Fam.Df$Family)[2] <- "Carcharhinidae" # 6 levels.

# Data frame.
gg.carch.fam <- data.frame(pc1 = Carch.scores[rownames(Carch.Fam.Df),1],
                           pc2 = Carch.scores[rownames(Carch.Fam.Df),2],
                           pc3 = Carch.scores[rownames(Carch.Fam.Df),3],
                           family = factor(Carch.Fam.Df$Family),
                           age = factor(Carch.Fam.Df$Age.2))

gg.carch.fam$age <- with(gg.carch.fam,factor(gg.carch.fam$age,levels = rev(bins)))
# Compute family-level means at each time-interval.
mean.carch.fam <- gg.carch.fam %>% 
  group_by(family, age) %>%
  dplyr::summarise(grp.mean = mean(pc1),
                   grp.median = median(pc1),
                   grp.mode = estimate_mode(pc1),na.rm = T)

# Plot.
# Scales varies across rows and have been set to reflect the full range of each family
ggplot(gg.carch.fam, aes(x = pc1,fill = family, colour = family)) +
  geom_histogram(bins = 30) +
  scale_y_continuous(labels = scales::number_format(accuracy = 1)) +
  scale_x_continuous(n.breaks = 4) +
  geom_vline(aes(xintercept = grp.mean), 
             linetype = "dashed", size = 0.4,
             color = "black", data = mean.carch.fam) + geom_vline(aes(xintercept = grp.mode), 
             linetype = "solid", size = 0.4,
             color = "#EFC000FF", data = mean.carch.fam) +
  scale_color_discrete_sequential(palette = "Red-Blue",nmax = 6) +
  scale_fill_discrete_sequential(palette = "Red-Blue",nmax = 6) +
  theme_grey() + xlab("Principal Component 1 (64.78%)") + ylab("Frequency") +
  theme(strip.text.x = element_text(size = 7, angle = 75, face = "bold"),
        strip.text.y = element_text(size = 7, face = "bold"),
        axis.title = element_text(color = "black", face = "bold", size = 7),
        strip.background = element_rect(colour = "black")) +
  theme(axis.text.x = element_text(size = 6),
        axis.text.y = element_text(size = 6),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        legend.position = "none") + facet_grid(age ~ family,scales = "free",space = "free_x")
```

### Family-level disparity
- **NB: Sample size <2 are not considered as part of this analysis**
- For Lamniformes: 14 out 16 families are retained (**Xiphodolamiidae**, **Eoptolamnidae** and **Cretoxyrhinidae ** are not included).
- For Carcharhiniformes: **Leptochariidae and Pseudotriakidae** are not considered as part of this analysis because of low sample sizes.
```{r warning=FALSE,cache=FALSE,eval=FALSE}
# 1. Subset data frame.
#    Start by creating a new variable that represent a match of families and ages.
#    Primitive solution, but it works.
Lamn.Fam.Df$famAge <- NULL
Lamn.Fam.Df$famAge <- as.factor(Lamn.Fam.Df$Family %% Lamn.Fam.Df$Age.2)
Reduced.Lamn.Df <- Lamn.Fam.Df %>% rownames_to_column('X') %>%
  group_by(famAge) %>% filter(n() >= 5) %>%
  column_to_rownames('X')

Reduced.Lamn.Df$Family <- factor(Reduced.Lamn.Df$Family)
Reduced.Lamn.Df$Age.2 <- factor(Reduced.Lamn.Df$Age.2)

# 2. Split data frame by Age.
split.ls <- split(Reduced.Lamn.Df, factor(Reduced.Lamn.Df$Age.2))

# 3. Partial disparity.
partial.disps <- vector(mode = "list", length = length(split.ls))
for(i in 1:length(split.ls)) {
  levs <- droplevels(split.ls[[i]][,11])
  gdf.split <- geomorph.data.frame(coords = GPA$coords[,,rownames(split.ls[[i]])], grp = levs)
  if(length(levels(levs)) < 2) disp <- morphol.disparity(coords ~ 1, data = gdf.split)
  else disp <- morphol.disparity(coords~1, groups = ~grp, data = gdf.split, partial = TRUE)
  partial.disps[[i]] <- disp
}

# 4. Repeat and do the same for Carcharhiniformes.
Carch.Fam.Df$famAge <- NULL
Carch.Fam.Df$famAge <- as.factor(Carch.Fam.Df$Family %% Carch.Fam.Df$Age.2)
Reduced.Carch.Df <- Carch.Fam.Df %>% rownames_to_column('X') %>%
  group_by(famAge) %>% filter(n() >= 5) %>%
    column_to_rownames('X')

Reduced.Carch.Df$Family <- factor(Reduced.Carch.Df$Family)
Reduced.Carch.Df$Age.2 <- factor(Reduced.Carch.Df$Age.2)

# 5. Split data frame by Age.
split.ls.c <- split(Reduced.Carch.Df, factor(Reduced.Carch.Df$Age.2))

# 6. Partial disparity.
c.partial.disps <- vector(mode = "list", length = length(split.ls.c))
for(i in 1:length(split.ls.c)) {
  levs <- droplevels(split.ls.c[[i]][,11])
  c.gdf.split <- geomorph.data.frame(coords = GPA$coords[,,rownames(split.ls.c[[i]])], grp = levs)
  if(length(levels(levs)) < 2) disp <- morphol.disparity(coords ~ 1, data = c.gdf.split)
  else disp <- morphol.disparity(coords~1, groups = ~grp, data = c.gdf.split, partial = TRUE)
  c.partial.disps[[i]] <- disp
}
```

```{r}
# 7. Import PD results.
L.Fam.Disparity <- read.xlsx(file = "Metadata/Family-level Disparity.xlsx", sheetIndex = 1)
C.Fam.Disparity <- read.xlsx(file = "Metadata/Family-level Disparity.xlsx", sheetIndex = 2)
# 8, Re-level.
L.Fam.Disparity$Time <- factor(x = L.Fam.Disparity$Time,levels = bins)
C.Fam.Disparity$Time <- factor(x = C.Fam.Disparity$Time,levels = bins)
# 9. Plot.
ggplot(data = L.Fam.Disparity, mapping = aes(x = Time, y = Disparity, fill = Family)) + 
  geom_bar(stat = "identity", color = "black") +
  coord_flip() + labs(title = "Lamniformes") + xlab("Age (Ma)") +
    scale_fill_manual(values = sequential_hcl(14, palette = "Reds 3"))+
  theme(axis.title = element_text(color = "#666666", face = "bold", size = 7),
        axis.text.y = element_text(angle = 0, size = 7),
        axis.text.x = element_text(size = 7),
        panel.spacing = unit(0, "lines"),legend.position = "right",aspect.ratio = 1.5/3)
# Plot.
ggplot(data = C.Fam.Disparity, mapping = aes(x = Time, y = Disparity, fill = Family)) + 
  geom_bar(stat = "identity",colour = "black") +
  coord_flip() + labs(title = "Carcharhiniformes") +
  scale_fill_manual(values = sequential_hcl(6, palette = "Blues 3")) +
  theme(axis.title = element_text(color = "#666666", face = "bold", size = 7),
        axis.text.y = element_text(angle = 0, size = 7),
        axis.text.x = element_text(size = 7),
        panel.spacing = unit(0, "lines"),legend.position = "right",aspect.ratio = 1.5/3)
```

### Family-level sample size
```{r}
# Plot.
LFS <- ggplot(L.Fam.Disparity, aes(x = Time, y = Sample, fill = Family)) + 
  geom_bar(stat = "identity") +
  scale_fill_manual(values = sequential_hcl(14, palette = "Reds 3"))+
  xlab("Age") + ylab("Families") +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 45,vjust = .8,hjust = .8,size = 7),
        axis.title = element_text(color = "#666666", face = "bold", size = 7),
        axis.text.y = element_text(angle = 0, size = 7),
        panel.spacing = unit(0, "lines"),
        aspect.ratio = 2/3)

CFS <- ggplot(C.Fam.Disparity, aes(x = Time, y = Sample, fill = Family)) + 
  geom_bar(stat = "identity") +
  scale_fill_manual(values = sequential_hcl(6, palette = "Blues 3"))+
  xlab("Age") + ylab("Families") +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 45,vjust = .8,hjust = .8,size = 7),
        axis.title = element_text(color = "#666666", face = "bold", size = 7),
        axis.text.y = element_text(angle = 0, size = 7),
        panel.spacing = unit(0, "lines"),
        aspect.ratio = 2/3)

grid.arrange(arrangeGrob(LFS,CFS,ncol = 1,nrow = 2))
```

### Epoch-based clade disparity analysis
- Disparity is calculated using a coarser time binning scheme.
- Time bins include: 1) Late Cretaceous 2) Paleocene 3) Eocene 4) Oligocene 5) Miocene 6) Pliocene 7) Pleistocene 8) Holocene.
- **NB:** For Lamniformes the Pleistocene (N=4) is omitted.
```{r message=FALSE,warning=FALSE,cache=FALSE,eval=FALSE}
# Drop level.
ex.Pleistocene <- Lamn.Df[Lamn.Df$Epoch != "Pleistocene", ]
ex.Pleistocene$Epoch <- factor(ex.Pleistocene$Epoch)
# Lamniformes.
L.Epoch.Disp <- error.plot(gpa.coords = GPA$coords[,,rownames(ex.Pleistocene)],
                        blank = FALSE,groups = ex.Pleistocene$Epoch,
                        order = c(7,5,1,4,3,6,2),
                        replicates = 999,rarefy.par = list(min.N = 55,reps = 999))
# Carcharhiniformes.
C.Epoch.Disp <- error.plot(gpa.coords = GPA$coords[,,rownames(Carch.Df)],
                        blank = FALSE,groups = Carch.Df$Epoch,
                        order = c(8,5,1,4,3,7,6,2),
                        replicates = 999,rarefy.par = list(min.N = 15,reps = 999))

# Re-sampling procedure.
disparity.calc(gpa = GPA,data = ex.Pleistocene,ages = "Epoch")$PV.dist.Pval
disparity.calc(gpa = GPA,data = Carch.Df,ages = "Epoch")$PV.dist.Pval
```

### Plot Disparity Results
```{r, warning=F, message=F}
Epoch.Res <- read.xlsx(file = "Metadata/Epoch Disparity.xlsx",sheetIndex = 1)
Epoch.Res$Stages <- factor(Epoch.Res$Stages, levels = c("Late Cretaceous","Paleocene","Eocene",
                                                        "Oligocene","Miocene","Pliocene","Pleistocene",
                                                        "Holocene"))

# Epoch.Res$Clade <- relevel(Epoch.Res$Clade ,"Lamniformes")
e.pd <- position_dodge(4)

# Plot disparity.
Ep.1 <- ggplot(Epoch.Res, aes(x = Midpoint, y = Disparity, colour = Clade)) +
  scale_x_reverse() +
  theme_grey() +
  geom_errorbar(aes(ymin = LowerBoot.PI ,ymax = UpperBoot.PI), width = 1.5, position = e.pd, show.legend = FALSE,size = .5) +
  geom_point(position = e.pd, size = 3, shape = 21, fill = "white") +
  scale_colour_manual(values = c("#E84646","#377EB8")) +
  ylab("Procrustes Variance") + xlab("") +
  theme(axis.title.x = element_text(face = "bold"),
        axis.title.y = element_text(face = "bold"),
        axis.text.y = element_text(angle = 90, size = 7),
        axis.text.x = element_text(size = 7),
        legend.position = "none",panel.spacing = unit(0, "lines"),
        aspect.ratio = 1)

# Plot sample size by epoch.
Ep.2 <- ggplot(Epoch.Res, aes(x = factor(Stages),y = N)) +
  geom_bar(stat="identity",aes(fill = Clade)) +
  theme_classic() + xlab("") + 
  theme(axis.text.y = element_text(angle = 90, vjust = 1, hjust = 1,size = 7),
        axis.text.x = element_text(angle = 45, size = 7),
        legend.position = "none",
        panel.spacing = unit(0, "lines"),
        aspect.ratio = 1/2) +
  scale_fill_manual(values = c("#E84646","#377EB8"))

grid.arrange(arrangeGrob(Ep.1),Ep.2,ncol = 2,widths = c(5,4))
```

### Extant-data preparation
- Extant species morphospace, disparity, and ecological analyses.
- See next section.
- **NB:** A total of 64 extant species with shape-data.
```{r warning=FALSE}
# Extract extant only entries.
recent <- "Recent"
recent.df <- Data[Data$Age.2 %in% recent, ] # 1380 specimens.
# New column variable matching genus and species names
Genus <- recent.df$Genus
Species <- recent.df$Species
'%%' <- function(x,y) {paste(x,y, sep = "_")}
recent.df$GenSpe <- NULL
recent.df$GenSpe <- as.factor(Genus %% Species)
levels(recent.df$GenSpe)[27] <- "Carcharhinus_seali"
```

### Extant species-level disparity
- Procrustes variance (PV) is estimated for all sampled species using their group means.
- Because variance is calculated on a species-level, the observed variance, correspond to intraspecific variation (but in one sense also heterodonty).
- The results from this section are used later for correlation analysis with dietary niche breadth.
- Print table results.
```{r cache=FALSE,eval=FALSE}
# Lamniformes.
Recent.Lamn.Df <- subset(x = recent.df,subset = recent.df$Order == "Lamniformes")
Lamn.Extant.Coord <- GPA$coords[,,rownames(Recent.Lamn.Df)]
# Procrustes variance.
spe.L.df <- geomorph.data.frame(shape = Lamn.Extant.Coord, gp = factor(Recent.Lamn.Df$Species))
lamn.species <- morphol.disparity(f1 = shape ~ gp,groups = ~gp,
                                  iter = 999, data = spe.L.df)

# Concatenate results into a data table.
L.Table <- matrix(NA, ncol = 11, nrow = 11, dimnames = dimnames(lamn.species$PV.dist.Pval))
L.Table[upper.tri(L.Table)] <- lamn.species$PV.dist.Pval[upper.tri(lamn.species$PV.dist.Pval)] 
L.Table[lower.tri(L.Table)] <- lamn.species$PV.dist[lower.tri(lamn.species$PV.dist)] 
# Carcharhiniformes (52 extant taxa)
Recent.Carch.Df <- subset(x = recent.df,subset = recent.df$Order == "Carcharhiniformes")
Carch.Extant.Coord <- GPA$coords[,,rownames(Recent.Carch.Df)]
levels(Recent.Carch.Df$Species)[43] <- "seali"
# Procrustes variance.
spe.C.df <- geomorph.data.frame(shape = Carch.Extant.Coord,gp = factor(Recent.Carch.Df$GenSpe))
carch.species <- morphol.disparity(f1 = shape ~ gp,groups = ~gp,iter = 999,
                                   data = spe.C.df)
# Concatenate into a data table.
c.Table <- matrix(NA, ncol = 52, nrow = 52, dimnames = dimnames(carch.species$PV.dist.Pval))
c.Table[upper.tri(c.Table)] <- carch.species$PV.dist.Pval[upper.tri(carch.species$PV.dist.Pval)] 
c.Table[lower.tri(c.Table)] <- carch.species$PV.dist[lower.tri(carch.species$PV.dist)] 
```

### Plot disparity barplots along with sample-sizes
- Intraspecific variability.
```{r cache=FALSE,eval=FALSE}
# Lamniformes.
z0 <- data.frame(lamn.species$Procrustes.var,Taxa = names(lamn.species$Procrustes.var))
colnames(z0)[1] <- "PV"
ggplot(z0,aes(Taxa,PV)) + geom_bar(stat = "identity", fill = "darkred") +
  coord_flip() + labs(title = "Lamniformes") +
  theme(axis.text.y = element_text(size = 7),
        axis.text.x = element_text(size = 7),
        panel.spacing = unit(0, "lines"),
        aspect.ratio = 1/2)

# Carcharhiniformes.
z1 <- data.frame(carch.species$Procrustes.var,Taxa = names(carch.species$Procrustes.var))
colnames(z1)[1] <- "PV"
ggplot(z1,aes(Taxa,PV)) + geom_bar(stat = "identity", fill = "darkblue") +
  coord_flip() + labs(title = "Carcharhiniformes") +
  theme(axis.title = element_text(size = 7),
        strip.text.x = element_text(size = 7, face = "bold"),
        axis.text.y = element_text(size = 7),
        axis.text.x = element_text(size = 7),
        panel.spacing = unit(0, "lines"),
        aspect.ratio = 2)
```

### Estimate species-level monognathic heterodonty
Monognathic heterodonty (on the lower and upper series) for 10 extant lamniform species is here estimated based on a complete tooth row of single individuals.
```{r cache=FALSE,eval=FALSE}
# Choose lamniform specimens.
sp.code <- c("LMNH 020305501.05","ROM R6339",
             "P.28129.001","LMNH 020105018.011","ROM R6833",
             "R6832","R8637","R7940","R8645","R8638")

L.monognathic.df <- Lamn.Df[Lamn.Df$Sp. %in% sp.code, ]
jaw.location <- split(L.monognathic.df,factor(L.monognathic.df$Dental.Unit))
# Teeth along the lower jaws.
lower.mono <- jaw.location[[1]]
# Teeth along the upper jaws.
upper.mono <- jaw.location[[2]]
# Get shape coordinates.
lower.coords <- GPA$coords[,,rownames(lower.mono)]
upper.coords <- GPA$coords[,,rownames(upper.mono)]
# GDF.
m1 <- geomorph.data.frame(coords = lower.coords,
                          unit = factor(lower.mono$Dental.Unit),
                          species = factor(lower.mono$Species))

m2 <- geomorph.data.frame(coords = upper.coords,
                          unit = factor(upper.mono$Dental.Unit),
                          species = factor(upper.mono$Species))

# Disparity for each dentition by species.
lower.heterodonty <- morphol.disparity(f1 = coords~species,groups = ~species,iter = 999,
                                      data = m1)
upper.heterodonty <- morphol.disparity(f1 = coords~species,groups = ~species,iter = 999,data = m2)
# Plot.
layout(matrix(1:2,2,1,byrow = T),respect = T)
barplot(lower.heterodonty$Procrustes.var,las=2,main = "Monognathic heterodonty on the lower jaw",
        cex.axis = .7,cex = .7,cex.main = .7)
barplot(upper.heterodonty$Procrustes.var,las=2,main = "Monognathic heterodonty on the upper jaw",
        cex.axis = .7,cex = .7,cex.main = .7)

dev.off()
```

### Extant clade-level disparity
- *Between the two clades which one is more morphologically diverse?*
- Report result in the main text.
```{r}
Extant.Shape <- GPA$coords[,,rownames(recent.df)]
# Procrustes variance.
clade.Df <- geomorph.data.frame(shape = Extant.Shape, gp = factor(recent.df$Order))
clade.Disp <- morphol.disparity(f1 = shape ~ gp,groups = ~gp,iter = 999,data = clade.Df)
# p.adjust(clade.Disp$PV.dist.Pval,method = "fdr") # a result that is also significant.
# Procrustes variance of clades but without filter-feeding species.
Exc.zoo <- which(recent.df$Genus == "Megachasma" | recent.df$Genus == "Cetorhinus"| recent.df$Genus == "Cetorhinus ")
ex.disp <- recent.df[-Exc.zoo, ]
ex.disp$Genus <- factor(ex.disp$Genus)
pv.ex <- geomorph.data.frame(coords = GPA$coords[,,rownames(ex.disp)],gp = factor(ex.disp$Order))
pv <- morphol.disparity(f1 = coords ~ gp,groups = ~gp,iter = 999,data = pv.ex)
```

### Transform landmark data array from specimen to species-level
- This is an essential step for subsequent comparative analyses.
- Here, species mean shape configurations are constructed for all 63 extant species.
```{r}
# Computing an average tooth shape/species
taxa <- levels(as.factor(recent.df$GenSpe))
sp.ave <- array(dim = c(160, 2, length(taxa)), dimnames = list(NULL, NULL, taxa))
for(i in 1:length(taxa)) {
  sp.teeth <- which(recent.df$GenSpe == taxa[i])
  teeth.names <- rownames(recent.df[sp.teeth,])
  if(length(sp.teeth) == 1) mean.shape <- GPA$coords[,,teeth.names]
  else mean.shape <- mshape(GPA$coords[,,teeth.names])
  sp.ave[,,i] <- mean.shape 
}

# Plot mean TPS-grids.
par(mfrow = c(10,10))
colnames <- dimnames(sp.ave)[[3]]
for (i in 1:63) {
x <- sp.ave[,,i]
tps_grid(x,x,grid.size = 1,shp = T,
         legend = F,shp.lwd = 6,legend.text = colnames[i]) 
}

dev.off()
```

### Average species-level PCA
```{r}
extant.pca <- plot(gm.prcomp(sp.ave),
                   cex = 1, pch = 19);title(main = "Extant species PCA")
text(extant.pca$PC.points[,1:2],labels = rownames(extant.pca$PC.points),cex = 0.3, offset = 3, col = "orange")
# picknplot.shape(x = extant.pca,methods = "points")

# Predict shape at extremes: replace [,1] for other axes.
preds <- shape.predictor(sp.ave, x = extant.pca$PC.points[,1],
                         Intercept = FALSE, 
                         pred1 = min(extant.pca$PC.points[,1]),
                         pred2 = max(extant.pca$PC.points[,1]),
                         pred3 = mean(extant.pca$PC.points[,1]))

layout(matrix(1:3,1,3,byrow = T),respect = F)
plotRefToTarget(mshape(sp.ave), preds$pred1);title("min")
plotRefToTarget(mshape(sp.ave), preds$pred2);title("max")
plotRefToTarget(mshape(sp.ave), preds$pred3);title("mean")
dev.off()
```

### Extant species-distributions
- Data from Ebert et al. 2013
```{r warning=FALSE}
species.Distr <- read.xlsx(file = "Metadata/Extant Species Database.xlsx",sheetIndex = 1)
# species.Distr$Order <- relevel(species.Distr$Order,"Lamniformes")
# Plot. 
ggplot(species.Distr, aes(x = Families, y = No..Species, fill = Families)) +
  geom_bar(stat = "identity") + 
  facet_wrap(.~Order,scales = "free",ncol = 1) + ylab("Number of valid species") +
  theme(axis.title = element_text(),
        axis.text.y = element_text(angle = 90, size = 7),
        axis.text.x = element_text(angle = 45,vjust = 0.5,hjust = .75, size = 7)) +
  theme(axis.line = element_line(linetype = "solid"),
        legend.position = "none",panel.spacing = unit(0, "lines"),
        aspect.ratio = 1)
```

### Trait exploration
```{r}
# Load traits dataset
traits.raw <- read.xlsx(file = "Metadata/Order Traits.xlsx",sheetIndex = 1)
# traits.raw$Order <- relevel(as.factor(traits.raw$Order),ref = "Lamniformes")
traits.raw$updated_iucn <- factor(traits.raw$updated_iucn,
                                  levels = c("NE","DD","LC","NT","VU","EN","CR"))

# Red List Status
ggplot(traits.raw,aes(x=updated_iucn,fill=as.factor(updated_iucn)))+
  xlab("IUCN Red List Status") +
  ylab("Species") +
  geom_bar() +
  scale_fill_manual("Legend",values = c("NE" = "grey47","DD" = "grey",
                                        "LC" = "forestgreen","NT" = "yellowgreen",
                                        "VU" = "gold","EN" = "orange","CR" = "red")) +
  facet_wrap(.~Order,scales = "free") +
  theme_grey()+
  theme(aspect.ratio = 1) +
  coord_flip()

# Body size 
ggplot(traits.raw,aes(y = size,x = Order,colour = factor(Order))) +
  geom_boxplot(outlier.colour = NULL,outlier.alpha = NULL,outlier.shape = 21) +
  ylab("Body size (TL)") +
  scale_fill_manual(values = c("#E84646","#377EB8",.5)) +
  scale_color_manual(values = c("#E84646","#377EB8",.5)) +
  theme(aspect.ratio = 1)
  
# Habitat
ggplot(data=subset(traits.raw, !is.na(habitat)), aes(x=habitat,fill = Order)) +
  geom_bar() +
  scale_fill_manual(values = c("#E84646","#377EB8",.5)) +
  scale_color_manual(values = c("#E84646","#377EB8",.5)) +
  facet_wrap(.~Order,scales = "free") +
  theme_grey()+
  theme(legend.position = "none", aspect.ratio = 1) + coord_flip()
    
# Vertical
ggplot(data=traits.raw, aes(x=vertical,fill = Order)) +
  geom_bar() +
  scale_fill_manual(values = c("#E84646","#377EB8",.5)) +
  scale_color_manual(values = c("#E84646","#377EB8",.5)) +
  facet_wrap(.~Order,scales = "free") +
  theme_grey()+
  theme(legend.position = "none",aspect.ratio = 1) + coord_flip()

# Thermoregulation
ggplot(data=traits.raw, aes(x=thermo,fill = Order)) +
  geom_bar() +
  scale_fill_manual(values = c("#E84646","#377EB8",.5)) +
  scale_color_manual(values = c("#E84646","#377EB8",.5)) +
  facet_wrap(.~Order,scales = "free") +
  theme_grey()+
  theme(legend.position = "none",aspect.ratio = 1) + coord_flip()
```

### Read in trophic dietary data
- Ecological data comes mostly from **Cortes (1999)**, but also from **Gonzalez-Pestana et al. (2018), Bizzarro et al. (2017), and Yano et al. (2007)**.
- Compositional (multivariate) diet data (see Aitchison 1986, for data-treatment).
- Rows are closures and diet-data are inherently non-normal (range from 0 to 1).
- For downstream analysis diet data was available for 55 species: 9 lamniform and 46 carcharhiniform species.
- NB: Diet data is arc-sin transformed prior to covariance analysis.
```{r}
diets <- read.xlsx(file = "Metadata/Taxon and Diet Data.xlsx",sheetIndex = 1)
rownames(diets) <- paste(diets$Genus, diets$Species, sep = "_")
```

### Prepare data for diet analysis
```{r}
# Match diet and coordinate data.
recent.coords <- intersect(rownames(diets),dimnames(sp.ave)[[3]])
sp.ave.X <- sp.ave[,,recent.coords] # 55 species.
# Diet variables and match with coordinate data.
diet.vars <- names(diets)[18:28]
diets <- diets[dimnames(sp.ave.X)[[3]],diet.vars]
# Coerce data frame to matrix: two.d.array() also works.
mx_diets <- data.matrix(diets[,1:11],rownames.force = TRUE)
# Apply arc-sin transformation for proportional/percentage data.
mx_diets <- asin(sqrt(mx_diets[,1:11]/100))
# Lamniformes diet.
lamn.diet <- mx_diets[c(47:55),]
# Lamniformes diet along with coords.
sp.ave.lamn <- intersect(rownames(lamn.diet),dimnames(sp.ave.X)[[3]])
sp.ave.lamn <- sp.ave.X[,,sp.ave.lamn]
# Carcharhiniformes diet
carch.diet <- mx_diets[-c(47:55),]
# Carcharhiniformes diet along with coords.
sp.ave.carch <- intersect(rownames(carch.diet),dimnames(sp.ave.X)[[3]])
sp.ave.carch <- sp.ave.X[,,sp.ave.carch]
```

### Two-block partial least squares analysis for Procrustes shape variables
- This is a non-parametric approach and makes use of RRPP for significance testing. Here we performed the analysis at four levels:
- a) both groups together.
- b) Lamniformes.
- c) Carcharhiniformes.
- d) Exclusion of filter-feeding species from Lamniformes.
- *How much of the variation in tooth morphology can be explained by diet?*
```{r}
# Both groups.
pls <- two.b.pls(A1 = sp.ave.X, A2 = mx_diets,iter = 999,print.progress = F)
summary(pls) # First result (a): r-PLS: 0.505, P-value: 0.014, Effect Size: 2.5179.

# Lamniformes.
pls.lamn <- two.b.pls(A1 = sp.ave.lamn, A2 = lamn.diet,iter = 999,print.progress = F)
summary(pls.lamn) # r-PLS: 0.862, P-value: 0.073, Effect Size: 1.7295.
                  # low sample size and hence a non-significant p-value.

# Carcharhiniformes.
pls.carch <- two.b.pls(A1 = sp.ave.carch, A2 = carch.diet,iter = 999,print.progress = F)
summary(pls.carch) # r-PLS: 0.584, P-value: 0.002, Effect Size: 3.5389.

# Sensitivity test.
# Macrophagous diet.
macro.diet <- mx_diets[c(47:48,50:52,54:55),]
# Macrophagous shapes.
sp.ave.macro <- intersect(rownames(macro.diet),dimnames(sp.ave.X)[[3]])
sp.ave.macro <- sp.ave.X[,,sp.ave.macro]
# PLS.
pls.macro <- two.b.pls(A1 = sp.ave.macro, A2 = macro.diet,iter = 999,print.progress = F)
summary(pls.macro) # r-PLS: 0.873, P-value: 0.09,Effect Size: 1.3829
layout(matrix(1:2,ncol = 2),respect = T)
plot(pls.macro)
text(pls.macro$XScores[,1],pls.macro$YScores[,1],labels = dimnames(sp.ave.macro)[[3]],
     cex = 0.7, offset = 2)
# Pie graph of corresponding diets.
barplot(pls.macro$right.pls.vectors[,1],las = 3)
# Shape extremes.
macro.shape <- shape.predictor(sp.ave.macro, macro.diet,Intercept = FALSE,method = "PLS",
                               pred1 = -0.25,pred2 = -0.10, pred3 = 0.05)

layout(matrix(1:3,1,3,byrow = T),respect = F)
plotRefToTarget(mshape(sp.ave.macro), macro.shape$pred1,mag = 4)
plotRefToTarget(mshape(sp.ave.macro), macro.shape$pred2,mag = 4)
plotRefToTarget(mshape(sp.ave.macro), macro.shape$pred3,mag = 4)
# Deformation grids using Thin Plate Splines
l.1 <- mshape(sp.ave.macro) + matrix(-0.25*(pls.macro$left.pls.vectors[,1]),byrow=T,160,2)
l.2 <- mshape(sp.ave.macro) + matrix(-0.10*(pls.macro$left.pls.vectors[,1]),byrow=T,160,2)
l.3 <- mshape(sp.ave.macro) + matrix(0.05*(pls.macro$left.pls.vectors[,1]),byrow=T,160,2)
# Plot.
layout(matrix(1:3,nrow = 1,ncol = 3,byrow = T),respect = T)
tps_grid(l.1,l.1,grid.size = 10,shp = T,legend = T,shp.lwd = 6)
tps_grid(l.2,l.2,grid.size = 10,shp = T,legend = T,shp.lwd = 6)
tps_grid(l.3,l.3,grid.size = 10,shp = T,legend = T,shp.lwd = 6)
```

### Plot results
```{r}
layout(matrix(c(1:6),ncol = 2, byrow = T),respect = F)
# PLS 1.
# a)
barplot(pls$right.pls.vectors[,1],las = 3,ylab = "Dietary Loadings (PLS1 Block2)",
        col = sequential_hcl(11,palette = "Grays"),border = NA,
        font.lab = 2,names.arg = colnames(mx_diets));title(main = "Dietary Singular Vector (PLS1 Block2)")

plot(pls,pch = 21, col = "black",bg = "grey75", cex = 1, font.lab = 2)
# b)
barplot(pls.lamn$right.pls.vectors[,1],las = 3,ylab = "Dietary Loadings (PLS1 Block2)",
        col = sequential_hcl(11,palette = "Reds 3"),border = NA,
        font.lab = 2,names.arg = colnames(mx_diets));title(main = "Dietary Singular Vector (PLS1 Block2)")
plot(pls.lamn,pch = 21, col = "black",bg = "grey75", cex = 1, font.lab = 2)
text(pls.lamn$XScores[,1],pls.lamn$YScores[,1],labels = rownames(lamn.diet),cex = 0.7, offset = 2)
# c)
barplot(pls.carch$right.pls.vectors[,1],las = 3,ylab = "Dietary Loadings (PLS1 Block2)",
        col = sequential_hcl(11,palette = "Blues 3"),border = NA,
        font.lab = 2,names.arg = colnames(mx_diets));title(main = "Dietary Singular Vector (PLS1 Block2)")
plot(pls.carch,pch = 21, col = "black",bg = "grey75", cex = 1, font.lab = 2)
text(pls.carch$XScores[,1],pls.carch$YScores[,1],labels = rownames(carch.diet),cex = 0.7, offset = 2)
abline(h = 0.0, col = "grey",lwd = 2)
dev.off()

# PLS2: Pearson correlation coefficient.
round(cor(pls$XScores[,2],pls$YScores[,2],method = "pearson"),3) # r-PLS is higher on PLS2.
# Lamn.
round(cor(pls.lamn$XScores[,2],pls.lamn$YScores[,2],method = "pearson"),3) # very high, but still non-significant.
# Carch.
round(cor(pls.carch$XScores[,2],pls.carch$YScores[,2],method = "pearson"),3) # slightly lower than PLS1.

# P and Z computations.
# apply(pls$XScores[,2,drop=F], 1, RRPP:::pval) # P-values
# apply(pls.lamn$XScores, 1, RRPP:::effect.size) # Z-scores

# Macrophagous lamniform sharks only.
layout(matrix(c(1:2),ncol = 2, byrow = T),respect = F)
barplot(pls.macro$right.pls.vectors[,1],las = 3,ylab = "Dietary Loadings (PLS1 Block2)",
        col = sequential_hcl(11,palette = "Reds 3"),border = NA,
        font.lab = 2,names.arg = colnames(mx_diets));title(main = "Dietary Singular Vector (PLS1 Block2)")
plot(pls.macro,pch = 21, col = "black",bg = "grey75", cex = 1, font.lab = 2)
text(pls.macro$XScores[,1],pls.macro$YScores[,1],labels = rownames(pls.macro$XScores),cex = 0.7, offset = 2)
```

### Plot PLS morphology
- Alternative way of getting shape coordinates at the extreme of PLS axes.
```{r TPS grids,warning=FALSE}
ref <- mshape(sp.ave)
shapes <- shape.names <- NULL
for (i in 1:ncol(pls$XScores)) {
        plsaxis.min <- min(pls$XScores[, i])
        plsaxis.max <- max(pls$XScores[, i])
        pls.min <- pls.max <- rep(0, dim(pls$XScores)[2])
        pls.min[i] <- plsaxis.min
        pls.max[i] <- plsaxis.max
        pls.min <- as.matrix(pls.min %*% (t(pls$left.pls.vectors))) + 
            as.vector(t(ref))
        pls.max <- as.matrix(pls.max %*% (t(pls$left.pls.vectors))) + 
            as.vector(t(ref))
        shapes <- rbind(shapes, pls.min, pls.max)
        shape.names <- c(shape.names, paste("PLS", i, "min", sep = ""), 
            paste("PLS", i, "max", sep = ""))
}

shapes <- arrayspecs(shapes, 160, 2)
shapes <- lapply(seq(dim(shapes)[3]), function(x) shapes[,, x])
names(shapes) <- shape.names

# Plot.
par(mfrow = c(3,2), mar = c(0,0,0,0))
plotRefToTarget(ref,shapes$PLS1min); plotRefToTarget(ref,shapes$PLS1max)
plotRefToTarget(ref,shapes$PLS2min); plotRefToTarget(ref,shapes$PLS2max)
plotRefToTarget(ref,shapes$PLS3min); plotRefToTarget(ref,shapes$PLS3max)
dev.off()
# Plot with Momocs.
par(mfrow = c(3,2), mar = c(0,0,0,0))
tps_grid(shapes$PLS1min,shapes$PLS1min,grid.size = 10,shp = T,legend = T,shp.lwd = 6)
tps_grid(shapes$PLS1max,shapes$PLS1max,grid.size = 10,shp = T,legend = T,shp.lwd = 6)
tps_grid(shapes$PLS2min,shapes$PLS2min,grid.size = 10,shp = T,legend = T,shp.lwd = 6)
tps_grid(shapes$PLS2max,shapes$PLS2max,grid.size = 10,shp = T,legend = T,shp.lwd = 6)
dev.off()
```

```{r}
# Deformation grids using Thin Plate Splines
c.1 <- mshape(sp.ave.carch) + matrix(-0.4784909*(pls.carch$left.pls.vectors[,1]),byrow=T,160,2)
c.2 <- mshape(sp.ave.carch) + matrix(0.1787440*(pls.carch$left.pls.vectors[,1]),byrow=T,160,2)
# Plot.
layout(matrix(1:2,nrow = 1,ncol = 2,byrow = T),respect = T)
tps_grid(c.1,c.1,grid.size = 10,shp = T,legend = T,shp.lwd = 6)
tps_grid(c.2,c.2,grid.size = 10,shp = T,legend = T,shp.lwd = 6)
```

### Comparative phylogenetic analysis (PCM)
- Tree from Stein et al. 2018 (downloaded via http://vertlife.org/sharktree/downloads/)
- Nexus file of the 1192 species fully resolved tree. It's rooted and includes branch lengths.
- Match tree with Procrustes shape variables.
- Between the shape data and the phylogeny a total of 59 species are available.
- Reason: Five species are missing in the phylogeny.
- Run analysis on 1) maximum clade credibility tree; and 2) first true tree from the list.
```{r}
# Distribution of trees: as multiphylo object.
phy.Tree <- ape::read.nexus(file = "Phylogeny/10.cal.tree.nex",force.multi = TRUE)
# RAxML tree.
RAxML <- read.nexus(file = "Phylogeny/610.tree.10Cal.RAxML.BS.nex")
```

### Compute a maximum clade credibility tree
```{r message=FALSE,warning=FALSE,cache=FALSE,eval=FALSE}
mcc.Tree <- maxCladeCred(phy.Tree,rooted = FALSE)
# Correct the spelling of C.seali in the tree.
mcc.Tree$tip.label[977] <- "Carcharhinus_seali"
is.ultrametric(mcc.Tree) # yes.
# Compare with the true topology.
co.obj <- cophylo(tr1 = phy.Tree[[1]],tr2 = mcc.Tree,rotate = T)
plot(co.obj,link.type = "curved",link.lwd = 1,link.lty = "solid",lwd = .3,
     link.col = make.transparent("#8B2252",0.25),fsize = .2)
```

### Tree topology for Lamniformes and Carcharhiniformes
```{r cache=FALSE,eval=FALSE}
t.obj <- mcc.Tree$tip.label[897:1192]
t.phy <- drop.tip(phy = mcc.Tree,tip = mcc.Tree$tip.label[-match(t.obj, mcc.Tree$tip.label)])
plot(t.phy, cex = .1);axisPhylo()
# Do the same, but from the RAxML tree.
lc.names <- RAxML$tip.label[445:610]
lc.drop <- drop.tip(phy = RAxML,tip = RAxML$tip.label[-match(lc.names, RAxML$tip.label)])
plot(lc.drop, cex = .4)
```

### Phylogenetic Generalized Least Squares (PGLS)
- PGLS regressions assume a Brownian Motion mode of evolution (=Markovian model).
- This test if there is a relationship between 'tooth shape' with 'factors' while accounting for phylogeny.
- *What species are missing in the tree?*
- Following taxa are not represented in the tree: "Carcharhinus_amboinensis" & "Carcharhinus_porosus"
- Shape = phylogeny = continuous variables.
- Subset the sp.ave.X object (coordinates).
- 55 taxa considered as part of this analysis.
```{r message=FALSE,warning=FALSE,cache=FALSE,eval=FALSE}
# Procrustes ANOVA within a phylogenetic framework.
# 1) Subset coordinates.
pgls.coords <- intersect(dimnames(sp.ave.X)[[3]],mcc.Tree$tip.label)
pgls.coords <- sp.ave.X[,,pgls.coords]
dim(pgls.coords) # 55 species.
# 2) Tree.
df.names <- data.frame(x = dimnames(pgls.coords)[[3]])
rownames(df.names) <- paste(df.names$x)
kk.x <- name.check(mcc.Tree,df.names)
pgls.TREE <- drop.tip(mcc.Tree,kk.x$tree_not_data)
# 3) Independent variables (diet) - arcsin transformed values.
pgls.diet <- mx_diets[dimnames(pgls.coords)[[3]],diet.vars]
# 4) Models.
pgls.x1 <- procD.pgls(f1 = pgls.coords ~ pgls.diet,phy = pgls.TREE, SS.type = "I",iter = 999)
anova(pgls.x1) # result is significant.
```

### Phylogenetic PLS
```{r message=FALSE,warning=FALSE,cache=FALSE,eval=FALSE}
phylo.integration(A = pgls.coords,A2 = pgls.diet,phy = pgls.TREE,iter = 999) 
# r-PLS: 0.577, P-value: 0.002; Effect size: 3.023
```

### Phylogenetic heat map
- Map diets on to the tree topology.
```{r message=FALSE,warning=FALSE,cache=FALSE,eval=FALSE}
tree.1 <- phy.Tree[[1]]
tree.1$tip.label[977] <- "Carcharhinus_seali"
heat.phy <- name.check(tree.1,df.names)
heat.phy.x <- drop.tip(tree.1,heat.phy$tree_not_data)
# Plot.
phylo.heatmap(pgls.TREE, pgls.diet, standardize = T,colors = sequential_hcl(100,palette = "Heat"))
```

```{r message=FALSE,warning=FALSE,cache=FALSE,eval=FALSE}
# PGLS Model using the first tree from the multip.phylo list.
pgls.x2 <- procD.pgls(f1 = pgls.coords ~ pgls.diet,phy = heat.phy.x, SS.type = "I",iter = 999)
anova(pgls.x2) # is significant.
```

### PGLS models on 200 phylogenetic trees
```{r message=FALSE,warning=FALSE,cache=FALSE,eval=FALSE}
respel.x <- phy.Tree[[]]$tip.label
respel.x[977] <- "Carcharhinus_seali"
# Rename tips in the tree.
Keep.Tree <- lapply(phy.Tree, rename.tips, respel.x)
"multiPhylo"-> class(Keep.Tree)

keep.sp <- pgls.TREE$tip.label
# Prune all trees to match diet data.
ntrees <- lapply(Keep.Tree,keep.tip,tip = keep.sp)
"multiPhylo"-> class(ntrees)

# Fit PGLS models.
nT <- 1:200
res <- matrix(NA,nrow = 200, ncol = 2); colnames(res) <- c("F-value", "P-value")
for(i in nT){
  pglsModel <- procD.pgls(f1 = pgls.coords ~ pgls.diet,phy = ntrees[[i]], SS.type = "I",iter = 999)
  res[i,1:2] <- c(summary(pglsModel)[[1]][1,5],summary(pglsModel)[[1]][1,7])
}

length(which(res[,2] < 0.05)) # 191 trees.
```

### Diet Breadth Analysis
- We compute a niche space [Ordinated diet niche breadth (ODB)] that correspond to various degrees of specialization.
- The data used here is a binary setup of Cortes 1999 diet composition data for a total of 100 species.
- Input data (presence/absence) is passed into a Jaccard dissimilarity matrix.
- *Underlying question: Do lamniform species have wider dietary niche breadth than carcharhiniforms?*
- We test if these clades and their constituent families are significantly different using a phylogenetic MANOVA.
- Finally, clade distributions are also compared using a K-S test.
```{r}
niche.breadth <- read.xlsx(file = "Metadata/Diet breadth Data.xlsx",sheetIndex = 1)
rownames(niche.breadth) <- paste(niche.breadth$Genus,niche.breadth$Species,sep = "_")
# Convert data frame into matrix.
niche.frame <- data.matrix(niche.breadth[,18:28],rownames.force = TRUE)
ordi.x <- ordi.breadth(niche.frame,dist.method = "jaccard")

names(ordi.x$species)
names(ordi.x$tot.breadth) <- ordi.x$species
names(ordi.x$scaled.breadth) <- ordi.x$species
# Coordinates for each diet category in PCoA space.
# Co-occuring items found in a sharks diet are closer in ordinated space. 
plot(ordi.x$plants.ord[,2]~ ordi.x$plants.ord[,1], data = ordi.x$plants.ord,
     pch = 19, col = "black", xlab = "Ordinated Diet Breadth PCoA 1",
     ylab = "Ordinated Diet Breadth PCoA 2", font.lab = 2,main = "PCoA space");abline(h = 0,v = 0,lty = 2)   
text(ordi.x$plants.ord[,1],ordi.x$plants.ord[,2],
     label = rownames(ordi.x$plants.ord), col = 'purple')

# K-S test.
ks.test(x = ordi.x$scaled.breadth[1:89],y = ordi.x$scaled.breadth[90:100])

# Data frame preparation.
Fm <- c(rep("Carcharhinidae",38),rep("Hemigaleidae",2),
        rep("Proscylliidae",1),rep("Pseudotriakidae",2),
        rep("Scyliorhinidae",21),rep("Sphyrnidae",6),
        rep("Triakidae",19),rep("Alopiidae",3),rep("Cetorhinidae",1),
        rep("Lamnidae",4),rep("Megachasmidae",1),rep("Carchariidae",1),rep("Mitsukurinidae",1))

nich.ggplot.df <- data.frame(axis = ordi.x$scaled.breadth, Index = 1:100,
                             species = ordi.x$species,
                             gp = c(rep("C",89),rep("L",11)),
                             fam.ob = Fm)

rownames(nich.ggplot.df) <- paste(nich.ggplot.df[,3])

# Main graph: Scaled ordinated diet breadth.
gg.1 <- ggplot(nich.ggplot.df,aes(x = Index, y = axis,color = gp)) + geom_point() +
  scale_color_manual(values = c("#377EB8","#E84646"))+
  scale_fill_manual(values = c("#377EB8","#E84646")) +
  geom_text(aes(label = species),hjust = 0, vjust = 0,size = 2.5) +
  labs(title = "Species ODB Range") + ylab("Scaled ordinated diet breadth") +
  theme(legend.position = "none",
        axis.title = element_text(size = 7),
        strip.text.x = element_text(size = 7, face = "bold"),
        axis.text.y = element_text(angle = 90, size = 7),
        axis.text.x = element_text(size = 7),
        panel.spacing = unit(0, "lines"),
        aspect.ratio = 1)
  
# Probability distribution histograms along with modal values.
# Lamniformes.
gg.2 <- ggplot(nich.ggplot.df, aes(x = axis, fill = gp)) +
        geom_density(alpha = .5) +
  scale_color_manual(values = c("#377EB8","#E84646"))+
  scale_fill_manual(values = c("#377EB8","#E84646")) +
   geom_vline(xintercept = c(0.25,0.1),size = 1, linetype = 1) +
    xlab("Scaled ordinated diet breadth") + ylab("Density") +
  theme(legend.position = "none",
        axis.title = element_text(size = 7),
        strip.text.x = element_text(size = 7, face = "bold"),
        axis.text.y = element_text(angle = 90, size = 7),
        axis.text.x = element_text(size = 7),
        panel.spacing = unit(0, "lines"),
        aspect.ratio = 1/4)

# Family-level distributions.
# Re-level.
nich.ggplot.df$fam.ob <- factor(nich.ggplot.df$fam.ob,
                                levels = c("Carcharhinidae","Hemigaleidae",
                                           "Proscylliidae","Pseudotriakidae","Scyliorhinidae",
                                           "Sphyrnidae","Triakidae","Alopiidae","Carchariidae",
                                           "Cetorhinidae","Lamnidae","Megachasmidae","Mitsukurinidae"))
# Graph.
gg.3 <- ggplot(nich.ggplot.df, aes(y = axis,x = fam.ob,fill = fam.ob)) +
  geom_boxplot(outlier.size = 1) +
  stat_summary(fun.data = mean_cl_boot, geom = "errorbar",width = .4, lwd = .5,
               linetype = 1, col = "black") +
  stat_summary(fun = "mean", geom = "point",lwd = 2,
               position = position_dodge(width = 1), color = "black", pch = 21) +
  theme(legend.position = "none",
        axis.title = element_text(size = 7),
        strip.text.x = element_text(size = 7, face = "bold"),
        axis.text.y = element_text(angle = 90, size = 7),
        axis.text.x = element_text(angle = 45,size = 7),
        panel.spacing = unit(0, "lines"),
        aspect.ratio = 1/4)
  
# Combine
grid.arrange(arrangeGrob(gg.1),gg.2,ncol = 2)
```

### Phylogenetic Anova and Post-hoc TukeyHSD
```{r warning=FALSE,cache=FALSE,eval=FALSE}
# Breadth analysis tree.
breadth.name <- name.check(mcc.Tree,nich.ggplot.df)
breadth.Tree <- drop.tip(mcc.Tree,breadth.name$tree_not_data)
# Trim ODB data frame to match tree.
phy.breadth <- nich.ggplot.df[breadth.Tree$tip.label,]

# Create variables and provide names for each.
# Grouping variable.

# a) Families
ord.phy.family <- as.factor(phy.breadth[,5])
names(ord.phy.family) <- rownames(phy.breadth)
# b) clades
ord.phy.clades <- as.factor(phy.breadth[,4])
names(ord.phy.clades) <- rownames(phy.breadth)
# Vector.
ord.phy.scores <- phy.breadth[,1]
names(ord.phy.scores) <- rownames(phy.breadth)

# Phylogenetic ANOVA
breadth.aov <- aov.phylo(ord.phy.scores~as.factor(ord.phy.family),phy = breadth.Tree, nsim = 999,test = "Pillai")
summary(aov(breadth.aov))

clade.breadth.aov <- aov.phylo(ord.phy.scores~as.factor(ord.phy.clades),phy = breadth.Tree, nsim = 999,test = "Pillai")
summary(aov(clade.breadth.aov))

# Tukey Honest Significant Differences.
Tukey.ODB.x <- TukeyHSD(aov(breadth.aov)) # No differences between groups.
Tukey.ODB.y <- TukeyHSD(aov(clade.breadth.aov)) # No difference.
# Plot
plot(Tukey.ODB.x, las = 2, col = "black",yaxt='n', ann=FALSE)
axis(1, at = 1:78, labels = FALSE, tck = -0.02)
text(y = 1:78, x = par()$usr[1] - 0.025*(par()$usr[2] - par()$usr[1]),
     labels=rownames(Tukey.ODB.x$group), srt = 0, adj = 1, 
     xpd = TRUE, cex = .5, las = 2)

# Save means.
means.Breadth <- aggregate(as.vector(ord.phy.scores)~as.factor(ord.phy.family),FUN = "mean")
```

### Niche breadth analysis of the Great white shark
- Ecological data from **Hussey et al. 2012**.
- Examine how the niche breadth of *C. carcharias* changes with ontogeny.
- Do they become more specialized with larger-size?
```{r warning=FALSE}
WhiteShark <- read.xlsx(file = "Metadata/White Shark Data (1).xlsx",sheetIndex = 1,header = T)
rownames(WhiteShark) <- WhiteShark[,1]
WhiteShark <- WhiteShark[,-1]

WhiteShark.NF <- as.matrix(x = WhiteShark)

# class(WhiteShark.NF)[[1]] <- "matrix"

# Convert into a matrix.
WhiteShark.NF <- data.matrix(frame = WhiteShark[,1:57],rownames.force = TRUE)
ordi.whiteshark <- ordi.breadth(dat = WhiteShark.NF,dist.method = "jaccard")

# ggplot data frame.
whiteShark.gg <- data.frame(axis = ordi.whiteshark$scaled.breadth, Index = 1:4,
                            size.classes = ordi.whiteshark$species)
# Plot.
ggplot(whiteShark.gg, aes(x = size.classes, y = axis)) +
  geom_segment(aes(x = size.classes, xend = size.classes, y = 0, yend = axis), color="grey") +
  geom_point(color = "black", size = 4) + labs(title = "White Shark Sampled from South African Waters") +
  xlab("Successive Size Classes") +
  ylab("Scaled ordinated diet breadth") +
  theme(axis.title = element_text(size = 7),
        strip.text.x = element_text(size = 7, face = "bold"),
        axis.text.y = element_text(angle = 90, size = 7),
        axis.text.x = element_text(angle = 45,size = 7,vjust = .5),
        panel.spacing = unit(0, "lines"),
        aspect.ratio = 1/3)
```

### Relative importance (%IRI) of functional prey categories
```{r}
IRI.df <- read.xlsx(file = "Metadata/White Shark Pie Graph Diet Data.xlsx",sheetIndex = 1)
pie.IRI <- ggplot(IRI.df,
                  aes(x = "", y = Relative.Importance...IRI., fill = Functional.Prey.Groups)) +
  geom_bar(stat = "identity") + theme_bw() + facet_wrap(.~Successive.Size.Classes,nrow = 1)
# Plot pie chart.
pie.IRI + coord_polar(theta = "y") + theme(axis.title = element_text(size = 7),
        strip.text.x = element_text(size = 7, face = "bold"),
        axis.text.y = element_text(size = 7),
        axis.text.x = element_text(size = 7),
        panel.spacing = unit(0, "lines"),
        aspect.ratio = 1)
```

### The relationship between ODB with extant tooth morphology
- 64 taxa with shapes.
- 100 taxa with niche breadth.
- Subset niche breadth to match shape-axis.
- Total of **55 species** with shape and niche breadth information.
```{r}
subset.obj <- intersect(ordi.x$species,dimnames(extant.pca$PC.points)[[1]])
# Run niche breadth analysis on species with morphological data.
NF <- niche.frame[subset.obj, ]
ord.NF <- ordi.breadth(NF,dist.method = "jaccard")
# Orders.
shark.orders <- factor(paste(Data$Genus, Data$Species, sep = "_"))
# Compute axis mean.
pc1.means <- tapply(PCA$x[which(Data$Age.2 == "Recent"),1], shark.orders[which(Data$Age.2 == "Recent")], mean)
pc2.means <- tapply(PCA$x[which(Data$Age.2 == "Recent"),2], shark.orders[which(Data$Age.2 == "Recent")], mean)
pc3.means <- tapply(PCA$x[which(Data$Age.2 == "Recent"),3], shark.orders[which(Data$Age.2 == "Recent")], mean)
# Frame.
pc.means <- data.frame(pc1.means, pc2.means, pc3.means)
pc.means <- na.omit(pc.means[rownames(pc.means) %in% rownames(diets),])
pc.means <- pc.means[rownames(diets),]

# Correlate shape-axis with niche-breadth axis.
summary(lm(ord.NF$scaled.breadth~pc.means[,1])) # No.
summary(lm(ord.NF$scaled.breadth~pc.means[,2])) # Yes.
summary(lm(ord.NF$scaled.breadth~pc.means[,3]))# No.

# Correlation
cor.test(ord.NF$scaled.breadth,pc.means[,2],method = "pearson")
```

### Exploring Bivariate Interpolation
```{r}
# PC1 vs PC2 vs ODB using cubic interpolation
shared <- ordi.x$species %in% rownames(pc.means)
rwb1 <- colorRampPalette(c("#8E063B", "white", "#023FA5"))
akima.int.a <- interp(pc.means[,1], pc.means[,2], ordi.x$scaled.breadth[shared], nx=250, ny=250, linear = F)
filled.contour(akima.int.a, color.palette = rwb1, xlab = "PC1", ylab = "PC2",
               plot.axes = {axis(1); axis(2); text(pc.means[,1:2],labels = rownames(pc.means), cex = 0.5)});title(main = "PC1 vs PC2 vs ODB using cubic interpolation",line = 1)
# PC1 vs PC3 vs ODB.
akima.int.b <- interp(pc.means[,1], pc.means[,3], ordi.x$scaled.breadth[shared], nx=250, ny=250, linear = F)
filled.contour(akima.int.b, color.palette = rwb1, xlab = "PC1", ylab = "PC3", plot.axes = {axis(1); axis(2); text(pc.means[,1],pc.means[,3], labels = rownames(pc.means), cex = 0.5)});title(main = "PC1 vs PC3 vs ODB using cubic interpolation",line = 1)
```

### Test the relationship between ODB and species disparity
- Is there a relationship between species disparity and ODB?
- Also perform a similar analysis but on lamniform monognathic heterodonty.
```{r}
# Data frames.
dispNich.Lamn <- read.xlsx(file = "Metadata/Disparity vs. ODB.xlsx",sheetIndex = 1)
dispNich.Carch <- read.xlsx(file = "Metadata/Disparity vs. ODB.xlsx",sheetIndex = 2)

# Adj.R, slope, intercept, and p-value.
summary(lm(dispNich.Lamn$PV ~ dispNich.Lamn$SNB)) # No.
summary(lm(dispNich.Lamn$PV ~ dispNich.Lamn$TB)) # No.
summary(lm(dispNich.Carch$PV ~ dispNich.Carch$SNB)) # No.
summary(lm(dispNich.Carch$PV ~ dispNich.Carch$TB)) # No.

# Pearson's product-moment correlation ...
cor.test(dispNich.Lamn$PV,dispNich.Lamn$SNB)

# Monognathic heterodonty vs. ODB
heterodonty.ODB.a <- read.xlsx(file = "Metadata/Heterodonty vs. ODB.xlsx",sheetIndex = 1)
heterodonty.ODB.b <- read.xlsx(file = "Metadata/Heterodonty vs. ODB.xlsx",sheetIndex = 2)
# Lower Model.
summary(lm(heterodonty.ODB.a$Monognathic.PV ~ heterodonty.ODB.a$SNB)) # No.
summary(lm(heterodonty.ODB.a$Monognathic.PV ~ heterodonty.ODB.a$TB)) # No.
# Upper Model.
summary(lm(heterodonty.ODB.b$Monognathic.PV ~ heterodonty.ODB.b$SNB)) # No.
summary(lm(heterodonty.ODB.b$Monognathic.PV ~ heterodonty.ODB.b$TB)) # No.
```

### Test the relationship between average shapes and ODB 
```{r cache=FALSE,eval=FALSE}
new.Frame <- geomorph.data.frame(shape = pgls.coords, ODB = ord.NF$scaled.breadth,
                                 phy = heat.phy.x)
new.Test <- procD.pgls(shape ~ ODB,phy = phy,iter = 999,data = new.Frame)
summary(new.Test)
```

### Plot linear model results
```{r warning=FALSE}
# Lamniformes.
plot_1 <- ggplot(dispNich.Lamn, aes(x = SNB, y = PV)) +
  xlab("Scaled niche breadth") + ylab("PV (Intraspecific heterodonty)") +
  geom_point() + 
  labs(title = "Disparity") +
  stat_smooth(method = "lm", col = "purple") +
  theme(axis.title = element_text(size = 7),
        strip.text.x = element_text(size = 7, face = "bold"),
        axis.text.y = element_text(angle = 90, size = 7),
        axis.text.x = element_text(size = 7))
# Carcharhiniformes
plot_2 <- ggplot(dispNich.Carch, aes(x = SNB, y = PV)) +
  xlab("Scaled niche breadth") + ylab("PV (Intraspecific heterodonty)") +
  geom_point() +
  stat_smooth(method = "lm", col = "purple") + 
  theme(axis.title = element_text(size = 7),
        strip.text.x = element_text(size = 7, face = "bold"),
        axis.text.y = element_text(angle = 90, size = 7),
        axis.text.x = element_text(size = 7))

# Shape vs. ODB.
DNB.Df <- data.frame(odb = ord.NF$scaled.breadth,
                     pc1 = pc.means[,1],pc2=pc.means[,2],
                     pc3=pc.means[,3])

plot_3 <- ggplot(DNB.Df,aes(odb,pc1)) + 
  geom_point() +
  labs(title = "Morphology") +
  xlab("Scaled niche breadth") + ylab("Morphology (PC1)") +
  stat_smooth(method = "lm", col = "darkgreen") +
  theme(axis.title = element_text(size = 7),
        strip.text.x = element_text(size = 7, face = "bold"),
        axis.text.y = element_text(angle = 90, size = 7),
        axis.text.x = element_text(size = 7))

plot_4 <- ggplot(DNB.Df,aes(odb,pc2)) + 
  geom_point() +
  xlab("Scaled niche breadth") + ylab("Morphology (PC2)") +
  stat_smooth(method = "lm", col = "darkgreen") +
  theme(axis.title = element_text(size = 7),
        strip.text.x = element_text(size = 7, face = "bold"),
        axis.text.y = element_text(angle = 90, size = 7),
        axis.text.x = element_text(size = 7))

plot_5 <- ggplot(DNB.Df,aes(odb,pc3)) + 
  geom_point() +
  xlab("Scaled niche breadth") + ylab("Morphology (PC3)") +
  stat_smooth(method = "lm", col = "darkgreen") +
  theme(axis.title = element_text(size = 7),
        strip.text.x = element_text(size = 7, face = "bold"),
        axis.text.y = element_text(angle = 90, size = 7),
        axis.text.x = element_text(size = 7))

# Combine plots.
grid.arrange(plot_1,plot_2,plot_3,plot_4,plot_5,
             layout_matrix = cbind(c(1,2,NA), c(3,4,5)))
# grid.arrange(plot_1,plot_2,plot_3,plot_4,plot_5,nrow = 3,ncol = 5)
```

### Additional exploration
- Do specimens taken in **labial versus lingual view** provide different signals in shape space and disparity?
- Total sample size: 578 specimens (Carcharhiniformes = 226, Lamniformes = 352).
```{r warning=FALSE,cache=FALSE,eval=FALSE}
# 1. Data frame.
ling.lab <- read.xlsx(file = "Metadata/Lingual vs. Labial Data.xlsx",sheetIndex = 1)
rownames(ling.lab) <- paste(ling.lab$File.Name,ling.lab$File.Type,sep = "")
# 2. Landmark.
p.Lms <- readland.tps(file = "Duplicates.tps",specID = "ID")
p.Lms <- p.Lms[,,intersect(x = rownames(ling.lab),y = dimnames(p.Lms)[[3]])]
sliders <- rbind(define.sliders(1:79), define.sliders(80:160)) 
# 3. GPA.
viewGPA <- gpagen(A = p.Lms,curves = as.matrix(sliders), ProcD = FALSE)
# 4. PCA.
pca <- gm.prcomp(viewGPA$coords)

# 5. Geomorph dataframe.
ll.gdf <- geomorph.data.frame(coords = viewGPA$coords,view = factor(ling.lab$View),
                              order = factor(ling.lab$Order))
# 6. Subset.
grp.c <- coords.subset(A = viewGPA$coords,group = factor(ling.lab$Order))
# 7. Lamniformes GDF: 176 vs. 176.
l.gdf <- geomorph.data.frame(coords = grp.c$Lamniformes,
                             order = factor(ling.lab$Order)[ling.lab$Order == "Lamniformes"],
                             view = factor(ling.lab$View)[ling.lab$Order == "Lamniformes"])
# 8. Carcharhiniformes GDF: 113 vs 113.
c.gdf <- geomorph.data.frame(coords = grp.c$Carcharhiniformes,
                             order = factor(ling.lab$Order)[ling.lab$Order == "Carcharhiniformes"],
                             view = factor(ling.lab$View)[ling.lab$Order == "Carcharhiniformes"])

# Disparity: general level.
a <- morphol.disparity(f1 = coords ~ 1,groups = ~view,iter = 999,data = ll.gdf)
# Disparity: specific level.
b <- morphol.disparity(f1 = coords ~ 1,groups = ~view,iter = 999,data = l.gdf)
f <- morphol.disparity(f1 = coords ~ 1,groups = ~view,iter = 999,data = c.gdf)
# Plot results.
layout(matrix(c(1,2,3),nrow = 1,ncol = 3),respect = T)
barplot(a$Procrustes.var,beside = T,ylab = "Procrustes Variance", main = "Combined")
barplot(b$Procrustes.var,main = "Lamniformes")
barplot(f$Procrustes.var, main = "Carcharhiniformes")
```

**Create variables for OLS analysis**
```{r warning=FALSE,cache=FALSE,eval=FALSE}
# Labial teeth.
pc1.labial <- pca$x[ling.lab$View == "labial",1]
pc2.labial <- pca$x[ling.lab$View == "labial",2]
# Lingual teeth.
pc1.lingual <- pca$x[ling.lab$View == "lingual",1]
pc2.lingual <- pca$x[ling.lab$View == "lingual",2]
# Data frame.
df.views <- data.frame(scores = pca$x,views = ling.lab$View)
pc.1.df.views <- data.frame(x = pc1.labial, y = pc1.lingual)
pc.2.df.views <- data.frame(x = pc2.labial, y = pc2.lingual)

# Linear model fit.
lmf <- lm(pc1.labial~pc1.lingual,pc.1.df.views)
coefplot(lmf)
d <- coef(summary(lmf))
knitr::kable(d)
rownames(d) <- c("$\\beta_0$", "$\\beta_1$")
colnames(d)[4] <- "$P(T > |t|)$"
knitr::kable(d, escape = FALSE)

# Regression plots.
ov.1 <- data.frame(data = pca$x,x = pc1.labial,
                   y = pc1.lingual,
                   order = ling.lab$Order)
# ov.1$order <- relevel(x = ov.1$order,ref = "Lamniformes")
ov.2 <- data.frame(data = pca$x,x = pc2.labial,
                   y = pc2.lingual,
                   order = ling.lab$Order)
# ov.2$order <- relevel(x = ov.2$order,ref = "Lamniformes")

# Facet OLS regression plot.
ggplot(ov.1, aes(x = x, y = y)) + 
  geom_point(pch = 19, alpha = .5) +
  xlab("Principal Component 1 (Labial)") +
  ylab("Principal Component 1 (Lingual)") +
  theme_bw() + theme(legend.position = "none") +
  geom_smooth(method=lm,se = F, color = 'darkgreen', lty = 1, size = 1) +
  geom_abline(intercept=0, slope=1,size=1/2) +
  stat_cor(label.y = 0.5) +
  stat_regline_equation(label.y = 0.2) +
  theme(axis.title.x = element_text(face = 'bold',size = 10),
        axis.title.y.left = element_text(face = 'bold'),
        legend.position="top") + facet_wrap(~ order, ncol = 2) +
  theme(strip.text.x = element_text(size =8, face="bold"))

ggplot(ov.2, aes(x = x, y = y)) + 
  geom_point(pch = 19, alpha = .5) +
  xlab("Principal Component 2 (Labial)") +
  ylab("Principal Component 2 (Lingual)") +
  theme_bw() + theme(legend.position = "none") +
  geom_smooth(method=lm,se = F, color = 'darkgreen') +
  geom_abline(intercept=0, slope=1,size=1/2) +
  stat_cor(label.y = 1) +
  stat_regline_equation(label.y = 1.5) +
  theme(axis.title.x = element_text(face = 'bold',size = 10),
        axis.title.y.left = element_text(face = 'bold'),
        legend.position="top") + facet_wrap(~ order, ncol = 2) +
  theme(strip.text.x = element_text(size =8, face="bold"))
```

### The End

***
# Session Information
```{r}
devtools::session_info()
```